---
title: "Untitled"
output: html_document
date: "2023-07-18"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384
mkdir FastQC_Results_BS384
```


```{bash}
<!-- cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384 -->

<!-- # Define the path to your data -->
<!-- datapath="/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs" -->

<!-- # Run FastQC on the specific files -->
<!-- for filename in $datapath/*_R1.fastq.gz $datapath/*_R2.fastq.gz -->
<!-- do -->
<!--   fastqc -o FastQC_Results_BS384/ $filename -->
<!-- done -->
```


```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384

# Define the path to your data
datapath="/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs"

# Run FastQC on the specific files
ls $datapath/*_R1.fastq.gz $datapath/*_R2.fastq.gz | parallel 'fastqc -o FastQC_Results_BS384/ {}'

```

```{bash}
conda activate multiqc
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384/FastQC_Results_BS384
multiqc .
```

Averge quality of first 10 sequecne and last 30 sequence is bit low so these can
be trimmed.
Since each amplicon length after PCR amplification is ~350 its seems okay to trim
these sequences as we will still have whole amplified region after forward and reverse
merge.


in multiqc report there are lot of adapters and some bias in the first 10 bases. I am thinking to not trim
the first 10 bases as the primers are specifically starting from the first base.
trim_galore automaticllay detects adapters, remove them,
and the automatically generates fastqc report which can be converted to multiqc report
in the next step.
https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md
```{bash}

conda activate trimgalore_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1

# Define the path to your data
datapath="/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs"
# Run trim_galore
ls $datapath/*_R1.fastq.gz $datapath/*_R2.fastq.gz | parallel 'trim_galore --quality 30  --gzip --output_dir trim_galore/ {}'
#ls $datapath/*_R2.fastq.gz | parallel 'trim_galore --quality 30  --gzip --output_dir trim_galore/ {}'

## This takes about 1 hour to run

## Do fastqc
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384

# Define the path to your data
trimpath="/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/trim_galore"
mkdir FastQC_Results_trim_galored_BS384
# Run FastQC on the specific files
ls $trimpath/*_R1_trimmed.fq.gz $trimpath/*_R2_trimmed.fq.gz | parallel 'fastqc -o FastQC_Results_trim_galored_BS384/ {}'
```

```{bash}

```


```{bash}
conda activate multiqc
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Fastqc/Fastqc_BS384/FastQC_Results_trim_galored_BS384
multiqc . --interactive
```



The reads obtained by trim_galore trimmed reads are not working with bwa while
the raw reads work fine, As the gatk pipeline has its own adapter removal there is
no need to use the trim_galore files so I am deleting them.

```{bash}
rm -r  /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/trim_galore/
```


