---
title: "Untitled"
output: html_document
date: "2023-07-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

#Genome reference FROM: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/

# Genome reference name: GCA_000001405.15_GRCh38_full_analysis_set.fna.gz
# Genome reference index file: GCA_000001405.15_GRCh38_full_analysis_set.fna.fai

#dbSNP files: one is from https://ftp.ncbi.nlm.nih.gov/snp/latest_release/VCF/
# the other is from GATK 
```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38
# human genome reference
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.gz
gunzip GCA_000001405.15_GRCh38_full_analysis_set.fna.gz

# bwa index files
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz
tar -xzf GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz
rm GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz

# hg38 index file
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.fai
#wget https://ftp.ncbi.nlm.nih.gov/snp/latest_release/VCF/GCF_000001405.25.gz

```

The below code is for using trim_glaore adapter removed reads

```{bash}
conda activate GATK_env
# Directory containing your fastq files
FASTQ_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/trim_galore

# Output directory
OUT_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384

# Path to your reference genome
REFERENCE=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna

# Number of threads
THREADS=45

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile
#bwa index -a bwtsw ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna.gz

cd $FASTQ_DIR
#for R1 in NS.2150.TSP*---TSP200X.BS*_R1_trimmed.fq.gz
for R1 in NS.2150.TSP*---TSP200X.BS_006_R1_trimmed.fq.gz NS.2150.TSP*---TSP200X.BS_008_R1_trimmed.fq.gz

do
    # Replace _R1.fastq with _R2.fastq to get the paired file
    R2=${R1/_R1_trimmed.fq.gz/_R2_trimmed.fq.gz}

    # Extract sample name
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.2150.TSP[0-9]*---TSP200X.//' -e 's/_R1_trimmed.fq.gz//')

    # Align with BWA-MEM
    bwa mem -M -t $THREADS $REFERENCE $R1 $R2 > $OUT_DIR/${SAMPLE_NAME}.sam


<!--     # Convert to BAM, sort, and index -->
<!--     samtools view -@ $THREADS -S -b $OUT_DIR/${SAMPLE_NAME}.sam > $OUT_DIR/${SAMPLE_NAME}.bam -->
<!--     samtools sort -@ $THREADS $OUT_DIR/${SAMPLE_NAME}.bam -o $OUT_DIR/${SAMPLE_NAME}_sorted.bam -->
<!--     samtools index $OUT_DIR/${SAMPLE_NAME}_sorted.bam -->
<!--     # Mark duplicates -->
<!--     _JAVA_OPTIONS="-Xmx90g" picard MarkDuplicates I=$OUT_DIR/${SAMPLE_NAME}_sorted.bam O=$OUT_DIR/${SAMPLE_NAME}_marked.bam M=$OUT_DIR/${SAMPLE_NAME}_metrics.txt -->

done
```


```{bash}
bwa mem -M -t $THREADS $REFERENCE \
NS.2150.TSP0375---TSP200X.BS_492_R1_trimmed.fq.gz NS.2150.TSP0375---TSP200X.BS_492_R2_trimmed.fq.gz > $OUT_DIR/"BS_492.sam"

# output file BS_492.sam is only 18kb in size which is not right
```




```{bash}
bwa mem -M -t $THREADS $REFERENCE \
/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs/NS.2150.TSP0001---TSP200X.BS_006_R1.fastq.gz \
/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs/NS.2150.TSP0001---TSP200X.BS_006_R2.fastq.gz \
> $OUT_DIR/"BS_006.sam"

```

The reads obtained by trim_galore trimmed reads are not working with bwa while
the raw reads work fine, As the gatk pipeline has its own adapter removal there is
no need to use the trim_galore files so I am deleting them.

Now I will do analysis from raw reads without trimming. The adapter will be marked
by picard tool.
For this, I need to created bam file from fastq to clean the reads as given here
https://gatk.broadinstitute.org/hc/en-us/articles/4403687183515--How-to-Generate-an-unmapped-BAM-from-FASTQ-or-aligned-BAM
```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384
mkdir FastqToSam_file
```


```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file
# Activate conda environment
conda activate GATK_env

# Directory containing your fastq files
FASTQ_DIR=/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs

# Output directory
OUT_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file/

# Number of threads
THREADS=45

cd $FASTQ_DIR

# Set Java options
export _JAVA_OPTIONS="-Xmx90g"

# Function to process files
process_files() {
    R1=$1
    R2=${R1/_R1.fastq.gz/_R2.fastq.gz}
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.2150.TSP[0-9]*---TSP200X.//' -e 's/_R1.fastq.gz//')
    OUTPUT_name=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file/${SAMPLE_NAME}_FastqToSam.bam
    picard FastqToSam \
        FASTQ=$R1 \
        FASTQ2=$R2 \
        OUTPUT=$OUTPUT_name \
        SAMPLE_NAME=${SAMPLE_NAME} \
        READ_GROUP_NAME=BS384_NovoSeq6000 \
        LIBRARY_NAME=Fluidigm \
        PLATFORM=illumina
}

# Export the function so it can be used by parallel
export -f process_files

# Find R1 files and pass them to parallel
find . -name "*R1.fastq.gz" | parallel -j $THREADS process_files
# try with two files
#echo -e "NS.2150.TSP0001---TSP200X.BS_006_R1.fastq.gz\nNS.2150.TSP0002---TSP200X.BS_008_R1.fastq.gz" | parallel -j $THREADS process_files

## takes about 1.5 hour
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file

# Number of threads
THREADS=45
# 
process_picard_RevertSam() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_FastqToSam.bam//')
# RevertSam process to be given to parallel processing
  picard RevertSam \
    I=${SAMPLE_NAME}_FastqToSam.bam \
    O=${SAMPLE_NAME}_RevertSam.bam \
    SANITIZE=true \
    MAX_DISCARD_FRACTION=0.005 \
    ATTRIBUTE_TO_CLEAR=XT \
    ATTRIBUTE_TO_CLEAR=XN \
    ATTRIBUTE_TO_CLEAR=AS \
    ATTRIBUTE_TO_CLEAR=OC \
    ATTRIBUTE_TO_CLEAR=OP \
    ATTRIBUTE_TO_CLEAR=XS \
    SORT_ORDER=queryname  \
    RESTORE_ORIGINAL_QUALITIES=true \
    REMOVE_DUPLICATE_INFORMATION=true \
    REMOVE_ALIGNMENT_INFORMATION=true
    
  
  
}
# Export the function so it can be used by parallel
export -f process_picard_RevertSam

# pass them to parallel
# try with one files
#find . -name "*006_FastqToSam.bam" | parallel -j $THREADS process_picard_RevertSam # for one file
find . -name "*_FastqToSam.bam"     | parallel -j $THREADS process_picard_RevertSam

# takes about 30 minutes

## remove the input files to save the space
  rm *_FastqToSam.bam
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file

# Number of threads
THREADS=45
# 
process_picard_MarkIlluminaAdapters() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_RevertSam.bam//')
  
  _JAVA_OPTIONS="-Xmx90g" picard MarkIlluminaAdapters \
    I=${SAMPLE_NAME}_RevertSam.bam \
    O=${SAMPLE_NAME}_markilluminaadapters.bam \
    M=${SAMPLE_NAME}_markilluminaadapters_metrics.txt
    

}
# Export the function so it can be used by parallel
export -f process_picard_MarkIlluminaAdapters

# pass them to parallel
# try with one files
#find . -name "*006_RevertSam.bam" | parallel -j $THREADS process_picard_MarkIlluminaAdapters # for one file
find . -name "*_RevertSam.bam"   | parallel -j $THREADS process_picard_MarkIlluminaAdapters


```

https://gatk.broadinstitute.org/hc/en-us/articles/360039568932--How-to-Map-and-clean-up-short-read-sequence-data-efficiently


```{bash}
# Activate conda environment
conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38
picard CreateSequenceDictionary \
    R=GCA_000001405.15_GRCh38_full_analysis_set.fna \
    O=GCA_000001405.15_GRCh38_full_analysis_set.dict
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file

# Function for processing each file
process_SamToFastq_bwa_MergeBamAlignment() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_markilluminaadapters.bam//')
  # Path to your reference genome
  REF="/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"
  TMP=~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384
  # Number of cores
  THREADS=4
  _JAVA_OPTIONS="-Xmx90g" picard SamToFastq \
  I=$file \
  FASTQ=${SAMPLE_NAME}_samtofastq_interleaved.fq \
  CLIPPING_ATTRIBUTE=XT \
  CLIPPING_ACTION=2 \
  INTERLEAVE=true \
  NON_PF=true
  
  bwa mem -M -t $THREADS \
  -p ${REF} \
  ${SAMPLE_NAME}_samtofastq_interleaved.fq > ${SAMPLE_NAME}_bwa_mem.sam
  
  _JAVA_OPTIONS="-Xmx90g" picard MergeBamAlignment \
  ALIGNED_BAM=${SAMPLE_NAME}_bwa_mem.sam \
  UNMAPPED_BAM=${SAMPLE_NAME}_RevertSam.bam \
  OUTPUT=${SAMPLE_NAME}_aligned.bam \
  REFERENCE_SEQUENCE=${REF} \
  CREATE_INDEX=true \
  ADD_MATE_CIGAR=true \
  CLIP_ADAPTERS=false \
  CLIP_OVERLAPPING_READS=true \
  INCLUDE_SECONDARY_ALIGNMENTS=true \
  MAX_INSERTIONS_OR_DELETIONS=-1 \
  PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
  ATTRIBUTES_TO_RETAIN=XS
  
  rm ${SAMPLE_NAME}_samtofastq_interleaved.fq
  rm ${SAMPLE_NAME}_bwa_mem.sam
}

export -f process_SamToFastq_bwa_MergeBamAlignment

# Run the function in parallel for all *.bam files
#ls *006_markilluminaadapters.bam  | parallel -j $THREADS process_SamToFastq_bwa_MergeBamAlignment
#ls *_markilluminaadapters.bam     | parallel -j 10 process_SamToFastq_bwa_MergeBamAlignment

# Find all .bam files
bam_files=(*_markilluminaadapters.bam)

# Process only those files for which .bai file is missing
for bam_file in "${bam_files[@]}"; do
  SAMPLE_NAME=$(echo $bam_file | sed -e 's/_markilluminaadapters.bam//')
  bai_file="${SAMPLE_NAME}_aligned.bai"
  if [[ ! -f "$bai_file" ]]; then
    # Run the function in background (& at the end of command)
    process_SamToFastq_bwa_MergeBamAlignment "$bam_file" &
    # Ensure that we do not spawn too many processes
    while (( $(jobs | wc -l) >= 10 )); do
      # If we have too many jobs, wait before starting new one
      sleep 1
    done
  fi
done
```

Now we need to clear some space but before doing that do qulaity check for .bam
files at different stages and then delete all the files that are not required for
the subsequent steps
```{bash}
# Wait for all background jobs to finish
wait
<!-- ## remove the input files to save the space -->
<!-- rm *_markilluminaadapters.bam -->
<!-- rm *_RevertSam.bam -->
```


```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/GATK/All_20180418.vcf.gz
gunzip All_20180418.vcf.gz
zcat All_20180418.vcf.gz | less #to see the content and the version
# this version in GRCH38.p7 so download it from GATK repository and see the version it should match with GRCh38 only
https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false
#click on .vcf file and select Public URL

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf
cat Homo_sapiens_assembly38.dbsnp138.vcf | less
Check the version

```


```{bash}
#Mark duplicates

# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile/GATK_BS384/FastqToSam_file

# Function for processing each file
process_MarkDuplicates() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')

picard MarkDuplicates \
INPUT=${SAMPLE_NAME}_aligned.bam \
OUTPUT=${SAMPLE_NAME}_marked_duplicates.bam \
M=${SAMPLE_NAME}_marked_dup_metrics.txt

}

export -f process_MarkDuplicates

# Run the function in parallel for all *.bam files
 ls *006_aligned.bam  | parallel -j 40 process_MarkDuplicates
#ls   *_aligned.bam   | parallel -j 40 process_MarkDuplicates
```


```{bash}
gatk BaseRecalibrator \
-R /media/prashen/Second_stor/Vivianne/T2R_sequences/December_2021/REFERENCE/GCA_000001405.15_GRCh38_full_analysis_set.fna \
-I $(basename $filename) \
--known-sites /media/prashen/Second_stor/Vivianne/T2R_sequences/December_2021/REFERENCE/GCF_000001405.39_GRCh38_dbSNP155_chr_sorted.vcf.gz \
--known-sites /media/prashen/Second_stor/Vivianne/T2R_sequences/December_2021/REFERENCE/Homo_sapiens_assembly38.known_indels.vcf.gz \
-O $(basename $filename _marked_duplicates.bam)_recal_data_sample.table \
--tmp-dir /tmp/tmp_Vivi

```
















