---
title: "Untitled"
output: html_document
date: "2023-08-18"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

# libraries
```{r}
library(here)
library(tidyverse)
```

## trial

## sample map file

```{bash}

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CHRIM_80_Analysis/CHRIM_80_genes_55/CHRIM_80_55_GATK/sorted_BQSR_RG_55genes_HaplotypeCaller

ls *_raw_variants.g.vcf.gz | \
awk -F'_' -v PWD="$(pwd)" '{sample=$1; gsub(/^S/, "", sample); printf "CHRIM_%02d\t%s/%s\n", sample, PWD, $0}' | \
sort -k1,1 > \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CHRIM_80_55_map.txt

## Combine these map files for all the 550 samples 
```

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/sorted_BQSR_RG_55genes_HaplotypeCaller

ls *_raw_variants.g.vcf.gz | \
awk -F'_' -v PWD="$(pwd)" '{sample=$1; gsub(/^S/, "", sample); printf "CIHR_%03d\t%s/%s\n", sample, PWD, $0}' | \
sort -k1,1 > \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CIHR_96_55_map.txt
```

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR/sorted_BQSR_RG_55genes_HaplotypeCaller

ls *_raw_variants.g.vcf.gz | \
awk -F'_' -v PWD="$(pwd)" '{
    if ($1 == "BS" && $2 == "Chrim") {
        prefix="CHRIM";
        sample=$3;
        tag="_ii";
    } else if ($1 == "BS" && $3 == "ii") {
        prefix="CIHR";
        sample=$2;
        tag="_ii";
    } else if ($1 == "BS") {
        prefix="CIHR";
        sample=$2;
        tag="";
    }
    printf "%s_%s%s\t%s/%s\n", prefix, sample, tag, PWD, $0
}' | \
sort -k1,1 > \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CIHR_384_55_map.txt
```


## append
```{bash}
cat \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CHRIM_80_55_map.txt \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CIHR_96_55_map.txt \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/CIHR_384_55_map.txt | \
sort -k1,1 > \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/Samples550_55_map.txt

```

## Remove duplicated samples and rename the remainig ones accordingly
CHRIM_25,   CHRIM_59,     CHRIM_77,     CHRIM_79,   CIHR_092,     CIHR_243,     CIHR_246,     CIHR_251
CHRIM_25_ii, CHRIM_59_ii, CHRIM_77_ii, CHRIM_79_ii, CIHR_092_ii,  CIHR_243_ii,  CIHR_246_ii,  CIHR_251_ii

after that change the entries with _ii
eg.. CHRIM_25_ii to CHRIM_25
```{r}
entries2reomve=c("CHRIM_25",   "CHRIM_59",     "CHRIM_77",     "CHRIM_79",   "CIHR_092",     "CIHR_243",     "CIHR_246",     "CIHR_251")
Samples550_55_map = read.table(here::here("Samples550_Analysis/Samples550_GATK/Samples550_55_map.txt"),
                               sep = "\t")
Samples550_55_map_2 = Samples550_55_map %>% 
  dplyr::filter(!V1 %in% entries2reomve) %>%
  # to to remove the _ii suffix from the duplicated samples
  mutate(V1 = str_replace(V1, "_ii$", ""))
  
Samples550_55_map_2
## check the sample name and file path if it corresponds to the new sequences, especially for the duplciated ones

write.table(Samples550_55_map_2,
            here("Samples550_Analysis/Samples550_GATK/Samples550_55_map_2.txt"),
           sep = "\t", row.names = F, col.names = F, quote = F)
```



## one-time only merged bedtools required by GenomicsDBImport 
```{bash}
<!-- bedtools merge \ -->
<!-- -i <(bedtools sort -i /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15_no_names.bed) > \ -->
<!-- /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15_no_names_BedtoolsMerged.bed -->

```


```{bash}
conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK
# delete folder as it will be created automatically
rm -r ./Samples550_genomicsdb_workspace

_JAVA_OPTIONS="-Xmx90g" /home/wasif_pclab/Programs_Softwares/gatk-4.3.0.0/gatk GenomicsDBImport \
    --genomicsdb-workspace-path ./Samples550_genomicsdb_workspace \
    --sample-name-map ./Samples550_55_map_2.txt \
    -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15_no_names_BedtoolsMerged.bed
```


```{bash}
conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK

_JAVA_OPTIONS="-Xmx90g" /home/wasif_pclab/Programs_Softwares/gatk-4.3.0.0/gatk GenotypeGVCFs \
-R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
-V gendb:///media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples550_Analysis/Samples550_GATK/Samples550_genomicsdb_workspace \
-O Samples550_55genes_GenotypeGVCFs.vcf.gz

gunzip -c \
Samples550_55genes_GenotypeGVCFs.vcf.gz > \
Samples550_55genes_GenotypeGVCFs.vcf

```

