---
title: "Untitled"
output: html_document
date: "2023-08-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots
```{bash}
mkdir -p /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

## path for replacement  
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile 
# to 
/media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile

```

```{bash}
cd /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
# Activate conda environment
conda activate GATK_env


# Directory containing your fastq files
cd /media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs

# Set Java options
export _JAVA_OPTIONS="-Xmx90g"

# Function to process files
process_files() {
    R1=$1
    R2=${R1/_R1.fastq.gz/_R2.fastq.gz}
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.2150.TSP[0-9]*---TSP200X.//' -e 's/_R1.fastq.gz//')
    OUTPUT_name=/media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/${SAMPLE_NAME}_FastqToSam.bam
    _JAVA_OPTIONS="-Xmx90g" picard FastqToSam \
        FASTQ=$R1 \
        FASTQ2=$R2 \
        OUTPUT=$OUTPUT_name \
        SAMPLE_NAME=${SAMPLE_NAME} \
        READ_GROUP_NAME=BS384_NovoSeq6000 \
        LIBRARY_NAME=Fluidigm \
        PLATFORM=illumina
}

# Export the function so it can be used by parallel
export -f process_files

# Find R1 files and pass them to parallel
find . -name "*R1.fastq.gz" | parallel -j 40 process_files
# try with two files
#echo -e "NS.2150.TSP0001---TSP200X.BS_006_R1.fastq.gz\nNS.2150.TSP0002---TSP200X.BS_008_R1.fastq.gz" | parallel -j $THREADS process_files

## takes about 1.5 hour
```

```{bash}
# Activate conda environment
conda activate GATK_env

cd /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# 
process_picard_MarkIlluminaAdapters() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_FastqToSam.bam//')
  
  _JAVA_OPTIONS="-Xmx90g" picard MarkIlluminaAdapters \
    I=${SAMPLE_NAME}_FastqToSam.bam \
    O=${SAMPLE_NAME}_markilluminaadapters.bam \
    M=${SAMPLE_NAME}_markilluminaadapters_metrics.txt
    

}
# Export the function so it can be used by parallel
export -f process_picard_MarkIlluminaAdapters

# pass them to parallel
# try with one files
#find . -name "*006_RevertSam.bam" | parallel -j $THREADS process_picard_MarkIlluminaAdapters # for one file
find . -name "*_FastqToSam.bam"   | parallel -j 40 process_picard_MarkIlluminaAdapters

```


```{bash}
cd /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
mkdir -p Fastqc_report
mkdir -p Aligned_bamqc
```

## ***at this stage previous files were transferred to Jin_SBG_BMG computer for faster analysis*** 
```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

# Function for processing each file
process_SamToFastq_bwa() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_markilluminaadapters.bam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  # Number of cores
  THREADS=2
  _JAVA_OPTIONS="-Xmx90g" picard SamToFastq \
      I=$1 \
      FASTQ=${SAMPLE_NAME}_samtofastq_interleaved.fq \
      CLIPPING_ATTRIBUTE=XT \
      CLIPPING_ACTION=2 \
      INTERLEAVE=true \
      NON_PF=true \
      TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
  
  /home/mwkhan/anaconda3/envs/fastqc/bin/fastqc \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq \
      -o ./Fastqc_report
  
  bwa mem -M -t $THREADS \
      -p ${REF} \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq > \
      ${SAMPLE_NAME}_bwa_mem.sam
  
  rm ${SAMPLE_NAME}_samtofastq_interleaved.fq
}

export -f process_SamToFastq_bwa

# Run the function in parallel for all *.bam files
#ls *S1_markilluminaadapters.bam  | parallel -j 10 process_SamToFastq_bwa
ls *_markilluminaadapters.bam     | parallel -j 10 process_SamToFastq_bwa
```

```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./Fastqc_report \
-o ./Fastqc_report
```

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK
rm *_markilluminaadapters.bam
rm *_markilluminaadapters_metrics.txt
```


```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

# Function for processing each file
process_MergeBamAlignment() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_bwa_mem.sam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  
  _JAVA_OPTIONS="-Xmx80g" picard MergeBamAlignment \
  ALIGNED_BAM=$1 \
  UNMAPPED_BAM=${SAMPLE_NAME}_FastqToSam.bam \
  OUTPUT=${SAMPLE_NAME}_aligned.bam \
  REFERENCE_SEQUENCE=${REF} \
  CREATE_INDEX=true \
  ADD_MATE_CIGAR=true \
  CLIP_ADAPTERS=false \
  CLIP_OVERLAPPING_READS=true \
  INCLUDE_SECONDARY_ALIGNMENTS=true \
  MAX_INSERTIONS_OR_DELETIONS=-1 \
  PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
  ATTRIBUTES_TO_RETAIN=XS \
  TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}

export -f process_MergeBamAlignment

ls *_bwa_mem.sam     | parallel -j 10 process_MergeBamAlignment
```



```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK
rm *_bwa_mem.sam
```


```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

# Function for processing each file
process_picard_SortSam() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')
  
  _JAVA_OPTIONS="-Xmx80g" picard SortSam \
      I=${SAMPLE_NAME}_aligned.bam \
      O=${SAMPLE_NAME}_sorted.bam \
      SORT_ORDER=coordinate \
      TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}
export -f process_picard_SortSam

# Run the function in parallel for all *.bam files
ls *_aligned.bam  | parallel -j 20 process_picard_SortSam
```


```{bash}
conda activate qualimap_env

cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

rm -r    ./sorted_bamqc/Qualimap
mkdir -p ./sorted_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 5 \
      -bam ${SAMPLE_NAME}_sorted.bam \
      -outdir ./sorted_bamqc/Qualimap/${SAMPLE_NAME}_sorted_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /home/mwkhan/Taste_Genes_Analysis/References_Seqs/bed_files/Merged_bed_hg38_Jan15.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted.bam     | parallel -j 6 process_Qualimap



cd /home/mwkhan/Taste_Genes_Analysis/Samples_384_genes_55/Samples_384_5_GATK

/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_bamqc/Qualimap \
-o ./sorted_bamqc/Qualimap
```


## sorted bam files are copied to VERBATIM_HD_PC as back up.
## Use these sam files from VERBATIM_HD_PC to generate sorted_BQSR files in the computer 

```{bash}
# Activate conda environment
conda activate GATK_env
cd /media/wasif_pclab/VERBATIM_HD_PC/Samples550_Analysis/Samples550_sorted_sam_backup/Samples384_genes55_sorted_sam

out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR
rm -r     ${out_dir}
mkdir -p  ${out_dir}

# Function for processing each file
process_gatk_BaseRecalibrator() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  
  out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR
  
  ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
  
  _JAVA_OPTIONS="-Xmx90g" gatk BaseRecalibrator \
      -R $ref_file \
      -I $1 \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table

  _JAVA_OPTIONS="-Xmx90g" gatk ApplyBQSR \
      -R $ref_file \
      -I $1 \
      --bqsr-recal-file ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_BQSR.bam
}

export -f process_gatk_BaseRecalibrator
ls   *_sorted.bam   | parallel -j 20 process_gatk_BaseRecalibrator
```


```{bash}
conda activate qualimap_env

unset DISPLAY
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR

rm -r    ./sorted_BQSR_bamqc/Qualimap
mkdir -p ./sorted_BQSR_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_BQSR.bam \
      -outdir ./sorted_BQSR_bamqc/Qualimap/${SAMPLE_NAME}_sorted_BQSR_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_BQSR.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/wasif_pclab/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_BQSR_bamqc/Qualimap \
-o ./sorted_BQSR_bamqc/Qualimap
```

```{bash }
# Activate conda environment
conda activate GATK_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR

# Function for processing each file
process_picard_AddOrReplaceReadGroups() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR.bam//')
  SAMPLE_number=$(echo $1 | sed -e 's/_sorted_BQSR.bam//')
  
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" picard AddOrReplaceReadGroups \
        I=$1 \
        O=${SAMPLE_NAME}_sorted_BQSR_RG_55genes.bam \
        RGSM=$SAMPLE_number \
        RGLB=lib1 \
        RGPL=illumina \
        RGPU=unit1
}

export -f process_picard_AddOrReplaceReadGroups
ls   *_sorted_BQSR.bam   | parallel -j 40 process_picard_AddOrReplaceReadGroups
```


```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR

rm *.bam.bai
rm *_sorted_recal_data.table
```



```{bash}
# Activate conda environment
conda activate GATK_env

cd        /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR/sorted_BQSR_RG_55genes_HaplotypeCaller

# Function for processing each file
process_gatk_HaplotypeCaller() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR_RG_55genes.bam//')
  
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" gatk HaplotypeCaller \
      -R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
      -I $1 \
      -O sorted_BQSR_RG_55genes_HaplotypeCaller/${SAMPLE_NAME}_raw_variants.g.vcf.gz \
      --assembly-region-padding 100 \
      -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed \
      -ERC GVCF
}

export -f process_gatk_HaplotypeCaller
ls    *_sorted_BQSR_RG_55genes.bam  | parallel -j 40 process_gatk_HaplotypeCaller
```


```{bash}
conda activate qualimap_env

unset DISPLAY
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_384_Analysis/CIHR_384_genes_55/CIHR_384_55_GATK/sorted_BQSR

rm -r    ./sorted_BQSR_RG_55genes_bamqc/Qualimap
mkdir -p ./sorted_BQSR_RG_55genes_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR_RG_55genes.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_BQSR_RG_55genes.bam \
      -outdir ./sorted_BQSR_RG_55genes_bamqc/Qualimap/${SAMPLE_NAME}_sorted_BQSR_RG_55genes_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_BQSR_RG_55genes.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/wasif_pclab/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
--cl-config '{max_table_rows: 10000}' \
./sorted_BQSR_RG_55genes_bamqc/Qualimap \
-o ./sorted_BQSR_RG_55genes_bamqc/Qualimap
```