---
title: "gatk till vcf filtering"
output: html_document
date: "2023-07-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{bash}
conda create -n GATK_env
conda activate GATK_env
conda install -c bioconda gatk4
conda install -c bioconda picard

```

#Genome reference FROM: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/

# Genome reference name: GCA_000001405.15_GRCh38_full_analysis_set.fna.gz
# Genome reference index file: GCA_000001405.15_GRCh38_full_analysis_set.fna.fai

#dbSNP files: one is from https://ftp.ncbi.nlm.nih.gov/snp/latest_release/VCF/
# the other is from GATK 
```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38
# human genome reference
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.gz
gunzip GCA_000001405.15_GRCh38_full_analysis_set.fna.gz

# bwa index files
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz
tar -xzf GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz
rm GCA_000001405.15_GRCh38_full_analysis_set.fna.bwa_index.tar.gz

# hg38 index file
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.fai
#wget https://ftp.ncbi.nlm.nih.gov/snp/latest_release/VCF/GCF_000001405.25.gz

```

The below code is for using trim_glaore adapter removed reads

```{bash}
conda activate GATK_env
# Directory containing your fastq files
FASTQ_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/trim_galore

# Output directory
OUT_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384

# Path to your reference genome
REFERENCE=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna

# Number of threads
THREADS=45

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/GATK_pipepile
#bwa index -a bwtsw ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna.gz

cd $FASTQ_DIR
#for R1 in NS.2150.TSP*---TSP200X.BS*_R1_trimmed.fq.gz
for R1 in NS.2150.TSP*---TSP200X.BS_006_R1_trimmed.fq.gz NS.2150.TSP*---TSP200X.BS_008_R1_trimmed.fq.gz

do
    # Replace _R1.fastq with _R2.fastq to get the paired file
    R2=${R1/_R1_trimmed.fq.gz/_R2_trimmed.fq.gz}

    # Extract sample name
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.2150.TSP[0-9]*---TSP200X.//' -e 's/_R1_trimmed.fq.gz//')

    # Align with BWA-MEM
    bwa mem -M -t $THREADS $REFERENCE $R1 $R2 > $OUT_DIR/${SAMPLE_NAME}.sam


<!--     # Convert to BAM, sort, and index -->
<!--     samtools view -@ $THREADS -S -b $OUT_DIR/${SAMPLE_NAME}.sam > $OUT_DIR/${SAMPLE_NAME}.bam -->
<!--     samtools sort -@ $THREADS $OUT_DIR/${SAMPLE_NAME}.bam -o $OUT_DIR/${SAMPLE_NAME}_sorted.bam -->
<!--     samtools index $OUT_DIR/${SAMPLE_NAME}_sorted.bam -->
<!--     # Mark duplicates -->
<!--     _JAVA_OPTIONS="-Xmx90g" picard MarkDuplicates I=$OUT_DIR/${SAMPLE_NAME}_sorted.bam O=$OUT_DIR/${SAMPLE_NAME}_marked.bam M=$OUT_DIR/${SAMPLE_NAME}_metrics.txt -->

done
```


```{bash}
bwa mem -M -t $THREADS $REFERENCE \
NS.2150.TSP0375---TSP200X.BS_492_R1_trimmed.fq.gz NS.2150.TSP0375---TSP200X.BS_492_R2_trimmed.fq.gz > $OUT_DIR/"BS_492.sam"

# output file BS_492.sam is only 18kb in size which is not right
```




```{bash}
bwa mem -M -t $THREADS $REFERENCE \
/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs/NS.2150.TSP0001---TSP200X.BS_006_R1.fastq.gz \
/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs/NS.2150.TSP0001---TSP200X.BS_006_R2.fastq.gz \
> $OUT_DIR/"BS_006.sam"

```

The reads obtained by trim_galore trimmed reads are not working with bwa while
the raw reads work fine, As the gatk pipeline has its own adapter removal there is
no need to use the trim_galore files so I am deleting them.

Now I will do analysis from raw reads without trimming. The adapter will be marked
by picard tool.
For this, I need to created bam file from fastq to clean the reads as given here
https://gatk.broadinstitute.org/hc/en-us/articles/4403687183515--How-to-Generate-an-unmapped-BAM-from-FASTQ-or-aligned-BAM
```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384
mkdir FastqToGVCF
```


```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
# Activate conda environment
conda activate GATK_env

# Directory containing your fastq files
FASTQ_DIR=/media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs

# Output directory
OUT_DIR=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/

# Number of threads
THREADS=45

cd $FASTQ_DIR

# Set Java options
export _JAVA_OPTIONS="-Xmx90g"

# Function to process files
process_files() {
    R1=$1
    R2=${R1/_R1.fastq.gz/_R2.fastq.gz}
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.2150.TSP[0-9]*---TSP200X.//' -e 's/_R1.fastq.gz//')
    OUTPUT_name=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/${SAMPLE_NAME}_FastqToSam.bam
    picard FastqToSam \
        FASTQ=$R1 \
        FASTQ2=$R2 \
        OUTPUT=$OUTPUT_name \
        SAMPLE_NAME=${SAMPLE_NAME} \
        READ_GROUP_NAME=BS384_NovoSeq6000 \
        LIBRARY_NAME=Fluidigm \
        PLATFORM=illumina
}

# Export the function so it can be used by parallel
export -f process_files

# Find R1 files and pass them to parallel
find . -name "*R1.fastq.gz" | parallel -j $THREADS process_files
# try with two files
#echo -e "NS.2150.TSP0001---TSP200X.BS_006_R1.fastq.gz\nNS.2150.TSP0002---TSP200X.BS_008_R1.fastq.gz" | parallel -j $THREADS process_files

## takes about 1.5 hour
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# Number of threads
THREADS=45
# 
process_picard_RevertSam() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_FastqToSam.bam//')
# RevertSam process to be given to parallel processing
  picard RevertSam \
    I=${SAMPLE_NAME}_FastqToSam.bam \
    O=${SAMPLE_NAME}_RevertSam.bam \
    SANITIZE=true \
    MAX_DISCARD_FRACTION=0.005 \
    ATTRIBUTE_TO_CLEAR=XT \
    ATTRIBUTE_TO_CLEAR=XN \
    ATTRIBUTE_TO_CLEAR=AS \
    ATTRIBUTE_TO_CLEAR=OC \
    ATTRIBUTE_TO_CLEAR=OP \
    ATTRIBUTE_TO_CLEAR=XS \
    SORT_ORDER=queryname  \
    RESTORE_ORIGINAL_QUALITIES=true \
    REMOVE_DUPLICATE_INFORMATION=true \
    REMOVE_ALIGNMENT_INFORMATION=true
    
  
  
}
# Export the function so it can be used by parallel
export -f process_picard_RevertSam

# pass them to parallel
# try with one files
#find . -name "*006_FastqToSam.bam" | parallel -j $THREADS process_picard_RevertSam # for one file
find . -name "*_FastqToSam.bam"     | parallel -j $THREADS process_picard_RevertSam

# takes about 30 minutes

## remove the input files to save the space
  rm *_FastqToSam.bam
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# Number of threads
THREADS=45
# 
process_picard_MarkIlluminaAdapters() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_RevertSam.bam//')
  
  _JAVA_OPTIONS="-Xmx90g" picard MarkIlluminaAdapters \
    I=${SAMPLE_NAME}_RevertSam.bam \
    O=${SAMPLE_NAME}_markilluminaadapters.bam \
    M=${SAMPLE_NAME}_markilluminaadapters_metrics.txt
    

}
# Export the function so it can be used by parallel
export -f process_picard_MarkIlluminaAdapters

# pass them to parallel
# try with one files
#find . -name "*006_RevertSam.bam" | parallel -j $THREADS process_picard_MarkIlluminaAdapters # for one file
find . -name "*_RevertSam.bam"   | parallel -j $THREADS process_picard_MarkIlluminaAdapters


```

https://gatk.broadinstitute.org/hc/en-us/articles/360039568932--How-to-Map-and-clean-up-short-read-sequence-data-efficiently


```{bash}
# Activate conda environment
conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38
picard CreateSequenceDictionary \
    R=GCA_000001405.15_GRCh38_full_analysis_set.fna \
    O=GCA_000001405.15_GRCh38_full_analysis_set.dict
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# Function for processing each file
process_SamToFastq_bwa_MergeBamAlignment() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_markilluminaadapters.bam//')
  # Path to your reference genome
  REF="/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"
  TMP=~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384
  # Number of cores
  THREADS=4
  _JAVA_OPTIONS="-Xmx90g" picard SamToFastq \
  I=$file \
  FASTQ=${SAMPLE_NAME}_samtofastq_interleaved.fq \
  CLIPPING_ATTRIBUTE=XT \
  CLIPPING_ACTION=2 \
  INTERLEAVE=true \
  NON_PF=true
  
  bwa mem -M -t $THREADS \
  -p ${REF} \
  ${SAMPLE_NAME}_samtofastq_interleaved.fq > ${SAMPLE_NAME}_bwa_mem.sam
  
  _JAVA_OPTIONS="-Xmx90g" picard MergeBamAlignment \
  ALIGNED_BAM=${SAMPLE_NAME}_bwa_mem.sam \
  UNMAPPED_BAM=${SAMPLE_NAME}_RevertSam.bam \
  OUTPUT=${SAMPLE_NAME}_aligned.bam \
  REFERENCE_SEQUENCE=${REF} \
  CREATE_INDEX=true \
  ADD_MATE_CIGAR=true \
  CLIP_ADAPTERS=false \
  CLIP_OVERLAPPING_READS=true \
  INCLUDE_SECONDARY_ALIGNMENTS=true \
  MAX_INSERTIONS_OR_DELETIONS=-1 \
  PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
  ATTRIBUTES_TO_RETAIN=XS
  
  rm ${SAMPLE_NAME}_samtofastq_interleaved.fq
  rm ${SAMPLE_NAME}_bwa_mem.sam
}

export -f process_SamToFastq_bwa_MergeBamAlignment

# Run the function in parallel for all *.bam files
#ls *006_markilluminaadapters.bam  | parallel -j $THREADS process_SamToFastq_bwa_MergeBamAlignment
#ls *_markilluminaadapters.bam     | parallel -j 10 process_SamToFastq_bwa_MergeBamAlignment

# Find all .bam files
bam_files=(*_markilluminaadapters.bam)

# Process only those files for which .bai file is missing
for bam_file in "${bam_files[@]}"; do
  SAMPLE_NAME=$(echo $bam_file | sed -e 's/_markilluminaadapters.bam//')
  bai_file="${SAMPLE_NAME}_aligned.bai"
  if [[ ! -f "$bai_file" ]]; then
    # Run the function in background (& at the end of command)
    process_SamToFastq_bwa_MergeBamAlignment "$bam_file" &
    # Ensure that we do not spawn too many processes
    while (( $(jobs | wc -l) >= 10 )); do
      # If we have too many jobs, wait before starting new one
      sleep 1
    done
  fi
done
```

Since the intermediate files are occupying a lot of space we need to delete them
but do Qualimap and multiqc before deleting them

```{bash}
conda activate GATK_env
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# make a new directory to save the qualimap files
aligned_bam_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/aligned_bam
rm -r ${aligned_bam_dir}
mkdir ${aligned_bam_dir}

# Function for processing each file
process_Qualimap_MultiQC_aligned_bam() {
    file=$1
    SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')
    qualimap bamqc \
    -nt 1 \
    -bam ${SAMPLE_NAME}_aligned.bam \
    --genome-gc-distr hg19 \
    -outdir /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/aligned_bam/${SAMPLE_NAME}_aligned
}
export -f process_Qualimap_MultiQC_aligned_bam
# Run the function in parallel for all *.bam files
#ls *006_aligned.bam *008_aligned.bam   | parallel -j 40 process_Qualimap_MultiQC_aligned_bam
ls   *_aligned.bam | parallel -j 40 process_Qualimap_MultiQC_aligned_bam
```


```{bash}
conda activate multiqc
multiqc \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/aligned_bam \
--outdir /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/aligned_bam
```
Now we need to clear some space but before doing that do qulaity check for .bam
files at different stages and then delete all the files that are not required for
the subsequent steps
```{bash}
# Wait for all background jobs to finish
wait

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
## remove the input files to save the space
rm *_FastqToSam.bam
rm *_RevertSam.bam
rm *_markilluminaadapters.bam
rm *_markilluminaadapters_metrics.txt
```

```{bash}
#Mark duplicates

# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# Function for processing each file
process_MarkDuplicates() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')

_JAVA_OPTIONS="-Xmx90g" picard MarkDuplicates \
INPUT=${SAMPLE_NAME}_aligned.bam \
OUTPUT=${SAMPLE_NAME}_marked_duplicates.bam \
M=${SAMPLE_NAME}_marked_dup_metrics.txt

}

export -f process_MarkDuplicates

# Run the function in parallel for all *.bam files
#ls *006_aligned.bam  | parallel -j 40 process_MarkDuplicates
ls   *_aligned.bam   | parallel -j 2 process_MarkDuplicates # use with one thread only this uses about 35 cores and 25gb of RAM when peaked
```

```{bash}
conda activate qualimap_env
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF 

# make a new directory to save the qualimap files
marked_duplicates_bam_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/marked_duplicates_bam
rm -r ${marked_duplicates_bam_dir}
mkdir ${marked_duplicates_bam_dir}

# Function for processing each file
process_Qualimap_MultiQC_marked_duplicates_bam() {
    file=$1
    SAMPLE_NAME=$(echo $1 | sed -e 's/_marked_duplicates.bam//')
    qualimap bamqc \
    -nt 1 \
    -bam ${SAMPLE_NAME}_marked_duplicates.bam \
    --genome-gc-distr hg19 \
    -outdir /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/marked_duplicates_bam/${SAMPLE_NAME}_marked_duplicates
}
export -f process_Qualimap_MultiQC_marked_duplicates_bam

#ls *006_marked_duplicates.bam *008_marked_duplicates.bam   | parallel -j 40 process_Qualimap_MultiQC_marked_duplicates_bam
ls   *_marked_duplicates.bam | parallel -j 20 process_Qualimap_MultiQC_marked_duplicates_bam

conda activate multiqc
multiqc \
/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/marked_duplicates_bam \
--outdir /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/MultiQC_Results/marked_duplicates_bam
```

```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF 
rm *_aligned.bam
rm *_aligned.bai
```

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/GATK/All_20180418.vcf.gz
gunzip All_20180418.vcf.gz
zcat All_20180418.vcf.gz | less #to see the content and the version
# this version in GRCH38.p7 so download it from GATK repository and see the version it should match with GRCh38 only
https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false
#click on .vcf file and select Public URL

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf
cat Homo_sapiens_assembly38.dbsnp138.vcf | less
Check the version

wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz
gunzip Homo_sapiens_assembly38.known_indels.vcf.gz
```

Index the variants files
```{bash}
conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
gatk IndexFeatureFile -F /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf
gatk IndexFeatureFile -F /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF

# Function for processing each file
process_gatk_BaseRecalibrator() {
  #file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_marked_duplicates.bam//')
  
  ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
  _JAVA_OPTIONS="-Xmx90g" gatk BaseRecalibrator \
  -R $ref_file \
  -I $1 \
  --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf \
  --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf \
  -O ${SAMPLE_NAME}_recal_data_sample.table
  
  _JAVA_OPTIONS="-Xmx90g" gatk ApplyBQSR \
  -R $ref_file \
  -I $1 \
  --bqsr-recal-file ${SAMPLE_NAME}_recal_data_sample.table \
  -O ${SAMPLE_NAME}_BQSR_dedup_sample_final.bam
}

export -f process_gatk_BaseRecalibrator

# Run the function in parallel for all *.bam files
#ls *006_marked_duplicates.bam *008_marked_duplicates.bam  | parallel -j 2 process_gatk_BaseRecalibrator
ls   *_marked_duplicates.bam                                | parallel -j 20 process_gatk_BaseRecalibrator
```

```{bash}
# Activate conda environment
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
mkdir ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
# Function for processing each file
process_gatk_HaplotypeCaller() {
  #file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_BQSR_dedup_sample_final.bam//')
  ref_file=
  _JAVA_OPTIONS="-Xmx90g" gatk HaplotypeCaller \
  -R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
  -I $1 \
  -O VCF/${SAMPLE_NAME}_raw_variants.g.vcf.gz \
  -bamout VCF/${SAMPLE_NAME}_reassamble_sample.out.bam \
  -ERC GVCF \
  --assembly-region-padding 100 \
  -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed
}

export -f process_gatk_HaplotypeCaller

# Run the function in parallel for all *.bam files
#ls *006_BQSR_dedup_sample_final.bam *008_BQSR_dedup_sample_final.bam  | parallel -j 20 process_gatk_HaplotypeCaller
ls    *_BQSR_dedup_sample_final.bam                                   | parallel -j 20 process_gatk_HaplotypeCaller
```


```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
for i in *.vcf.gz; do echo `bcftools query -l $i`;echo $i;done | paste - - >  ./Samples384.map
```

```{bash}
<!-- conda activate GATK_env -->

<!-- cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF -->

<!-- gatk GenomicsDBImport \ -->
<!-- --sample-name-map ./Combined384_samples.map \ -->
<!-- -R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \ -->
<!-- -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed -->
<!-- --genomicsdb-workspace-path ./Combined384_all_database \ -->
<!-- --interval-padding 100 -->

# I think -L option required non overlapping positions while our bed file has some
# overlappin positions so this is not working



# fix this command
```

```{bash}
conda activate GATK_env

cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
# Start the CombineGVCFs command
cmd='_JAVA_OPTIONS="-Xmx90g" gatk CombineGVCFs -R ${ref_file} '
# provide reference file
cmd+= " -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed"
# Add the output argument
cmd+=" -O Samples384_Combined.g.vcf"
# Add each GVCF file to the command
for file in $(ls *_raw_variants.g.vcf.gz); do
  cmd+=" -V $file "
done

# Run the command
eval $cmd
```



```{bash}
#to see the sample names and sample counts in the combined vcf file
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
grep '^#CHROM' Samples384_Combined.g.vcf | cut -f10- 
grep '^#CHROM' Samples384_Combined.g.vcf | cut -f10- | tr '\t' '\n' | wc -l
# 
grep -v '^#' Samples384_Combined.g.vcf | wc -l
```


```{bash}
conda activate GATK_env

cd  ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF

_JAVA_OPTIONS="-Xmx90g" gatk GenotypeGVCFs \
-R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
-V ./Samples384_Combined.g.vcf \
-O ./Samples384_Combined_Genotyped.g.vcf
```

```{bash}
cd ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
echo "sample numbers"
grep '^#CHROM' Samples384_Combined.g.vcf | cut -f10- | tr '\t' '\n' | wc -l
echo "chromosomes in the vcf file"
grep -v '^#' ./Samples384_Combined_Genotyped.g.vcf | cut -f1 | sort | uniq
```


## apply variantrecalibrator and VQSR
download files required for that
https://gatk.broadinstitute.org/hc/en-us/articles/360036510892-VariantRecalibrator
https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?prefix=&forceOnObjectsSortingFiltering=false
```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/1000G_omni2.5.hg38.vcf.gz
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/1000G_omni2.5.hg38.vcf.gz.tbi

wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/1000G_phase1.snps.high_confidence.hg38.vcf.gz
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/1000G_phase1.snps.high_confidence.hg38.vcf.gz.tbi

wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/hapmap_3.3.hg38.vcf.gz
wget https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/hapmap_3.3.hg38.vcf.gz.tbi
```


```{bash}
<!-- conda activate GATK_env -->

<!-- cd  ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF -->

<!-- ref_folder=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs -->
<!--  _JAVA_OPTIONS="-Xmx90g" gatk VariantRecalibrator \ -->
<!--    -R ${ref_folder}/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \ -->
<!--    -V ./Samples384_Combined_Genotyped.g.vcf \ -->
<!--    --resource hapmap,known=false,training=true,truth=true,prior=15.0:${ref_folder}/hapmap_3.3.hg38.vcf.gz \ -->
<!--    --resource omni,known=false,training=true,truth=false,prior=12.0:${ref_folder}/1000G_omni2.5.hg38.vcf.gz \ -->
<!--    --resource 1000G,known=false,training=true,truth=false,prior=10.0:${ref_folder}/1000G_phase1.snps.high_confidence.hg38.vcf.gz \ -->
<!--    --resource dbsnp,known=true,training=false,truth=false,prior=2.0:${ref_folder}/Homo_sapiens_assembly38.dbsnp138.vcf \ -->
<!--    -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed \ -->
<!--    -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \ -->
<!--    -mode SNP \ -->
<!--    -O VariantRecalibrator_output.recal \ -->
<!--    --tranches-file VariantRecalibrator_output.tranches \ -->
<!--    --rscript-file VariantRecalibrator_output.plots.R -->
```

## the trances file is empty so gatk ApplyVQSR will not work and will give an error
the reason may be :
Insufficient data: VQSR requires a large amount of data to generate the model effectively.
If you're using a small number of samples or limited annotations, there may not be enough data for VQSR to work properly.

```{bash}
<!-- conda activate GATK_env -->
<!-- cd  ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF -->
<!-- ref_folder=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs -->

<!--  _JAVA_OPTIONS="-Xmx90g" gatk ApplyVQSR \ -->
<!--    -R ${ref_folder}/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \ -->
<!--    -V ./Samples384_Combined_Genotyped.g.vcf \ -->
<!--    -O ./Samples384_Combined_Genotyped_VQSR.g.vcf \ -->
<!--    --truth-sensitivity-filter-level 99.0 \ -->
<!--    --tranches-file VariantRecalibrator_output.tranches \ -->
<!--    --recal-file VariantRecalibrator_output.recal \ -->
<!--    -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed \ -->
<!--    -mode SNP -->
   
# this command wont work so we can not apply VQSR
```

## Now we need to do the hard filtering
```{bash}
<!-- conda activate GATK_env -->
<!-- cd  ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF -->
<!-- ref=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna -->

<!-- _JAVA_OPTIONS="-Xmx90g" gatk VariantFiltration \ -->
<!--     -R ${ref} \ -->
<!--     -V ./Samples384_Combined_Genotyped.g.vcf \ -->
<!--     --filter-expression "QD < 2.0 || FS > 60.0 || MQ < 40.0" \ -->
<!--     --filter-name "basic_filters" \ -->
<!--     -O ./Samples384_Combined_Genotyped_filtered.g.vcf -->

# we do not need to do this step becuase we will usse filtering using vcftools


```


# Remove samples from vcf files
As We do not have metadata for these samples

```{bash}
cd  ~/media_Second_stor_wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF/VCF
# ./BS_106,./BS_Chrim_25,./BS_Chrim_59,./BS_Chrim_77,./BS_Chrim_79
bcftools view \
-s ^./BS_103,./BS_Chrim_25,./BS_Chrim_59,./BS_Chrim_77,./BS_Chrim_79 \
./Samples384_Combined_Genotyped.g.vcf > \
./Samples379_Combined_Genotyped.g.vcf
```










