---
title: "Untitled"
output: html_document
date: "2024-04-30"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## R Markdown
```{r}
rm(list = ls())
```

# Load libraries
```{r}
library(VariantAnnotation)
library(tidyverse)
library(here)
library(ComplexHeatmap)
library(circlize)
library(tidyr)
library(RColorBrewer)
```
# create plink files

```{bash, eval=FALSE}
conda activate Plink_env
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_A06_mbQTL
cd      /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_A06_mbQTL

metadata_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Metadata538_files/
inupt_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK/

plink \
--vcf ${inupt_dir}/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno_IDcomplete2.vcf \
--double-id \
--make-bed \
--keep-allele-order \
--make-pheno ${metadata_dir}/Samples538_Combined_Phenotype.txt 1 \
--update-sex ${metadata_dir}/Samples538_Combined_Phenotype_with_sex_info.txt \
--recode \
--out ./Samples538_55gene_Plink
```


# Plink recode vcf
```{bash, eval=FALSE}
plink --file ./Samples538_55gene_Plink \
--recode A \
--out ./Samples538_55gene_Plink.vcf
```


# plink logistic association

```{bash, eval=FALSE}
conda activate Plink_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_A06_mbQTL
metadata_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Metadata538_files/


plink \
--file ./Samples538_55gene_Plink \
--geno 0.5 \
--maf 0.01 \
--not-chr x \
--hwe 0.001 \
--lasso 0.5 \
--ci 0.95 \
--recode A \
--covar ${metadata_dir}/Samples538_Combined_covariates.txt \
--covar-name Age,Sex,SEFI_score \
--allow-no-sex \
--out ./Samples538_55gene_Plink_filtered_geno.5_maf.01_hwe.001
```



BH correction in the assoc.logistic file
## Assign gene name to assoc file

```{r}
assoc_file = read.table(here("Samples538_Analysis/Samples538_A06_mbQTL/Samples538_55gene_Plink_filtered_geno.5_maf.01_hwe.001.lasso"),
                        header = T)
assoc_file = assoc_file %>% 
  mutate(CHR = as.character(CHR)) 
assoc_file

assoc_file_2 = assoc_file %>% 
  mutate(SNP_A1 = paste(SNP, A1, sep = "_"))
assoc_file_2
```


## load recoded vcf file to make a new vcf which as only the above variants
```{r}
Samples538_recoded_vcf = read.table(here("Samples538_Analysis/Samples538_A06_mbQTL/Samples538_55gene_Plink_filtered_geno.5_maf.01_hwe.001.raw"),
                        header = T)
Samples538_recoded_vcf

Samples538_recoded_vcf_2 = Samples538_recoded_vcf %>%
  column_to_rownames(var = "IID") %>%
  # select only the variants in the assoc file
  dplyr::select(assoc_file_2$SNP_A1) %>% 
  # replace missing values with 0
  mutate(across(everything(), ~if_else(is.na(.), 0, .)))
Samples538_recoded_vcf_2
```


## CLR normalized microbiome file with only significant species

```{r}

Maaslin2_16S_ITS_features = read.table(here("Samples538_Analysis/Microbiome538_files/Maaslin2_16S_ITS_DA_normalized_features.txt"),
                                       header = T)
Maaslin2_16S_ITS_features

Maaslin2_16S_ITS_features_2 = Maaslin2_16S_ITS_features %>%
  column_to_rownames(var = "IID") %>%
  dplyr::select(-FID)
Maaslin2_16S_ITS_features_2
```


##covariate

```{r}
# from another R project
metadata538 = read.csv(here("../RP_230728_Metadata_ECC_study/Samples538_combined/Samples538_metadata_17_columns_240306.txt"),
                        sep = "\t", header = TRUE, stringsAsFactors = FALSE)
metadata538
```


```{r}
metadata538 %>% group_by(Sex) %>% summarise(n = n())
metadata538 %>% group_by(ECC_status) %>% summarise(n = n())
```


Required by plink Binary ('0' = control, '1' = case)
Required by plink Sex code ('1' = male, '2' = female, '0' = unknown)
FID (Family ID) and IID (Individual ID) for it to correctly identify individuals

```{r}
metadata538_2 = metadata538 %>% 
  dplyr::select(sampleid, ECC_status, Sex, Age, Urban_status, SEFI_score) %>% 
  dplyr::rename(IID  = sampleid) %>%
  mutate(FID = IID) %>%
  mutate(Sex = case_when( # AS Original in data 1, Female | 2, Male 
    Sex == 0 ~ "Female",
    Sex == 1 ~ "Male")) %>% 
  mutate(Sex = case_when( # AS 	required by plink Sex code ('1' = male, '2' = female, '0' = unknown)
    Sex == "Female" ~ 2,
    Sex == "Male" ~ 1,
    TRUE ~ 0 )) %>% 
  dplyr::select(IID, Sex, Age, Urban_status) %>%
  column_to_rownames(var = "IID")

metadata538_2

metadata538_2_t = metadata538_2 %>%
  as.matrix() %>% 
  t() %>% 
  as.data.frame()
metadata538_2_t
```

```{r}
Samples538_LinearAnalysisTaxaSNP <- linearTaxaSnp(Maaslin2_16S_ITS_features_2,
  Samples538_recoded_vcf_2,
  metadata538_2_t
)
Samples538_LinearAnalysisTaxaSNP
```

```{r}
Samples538_LinearAnalysisTaxaSNP_2 = Samples538_LinearAnalysisTaxaSNP %>%
  dplyr::select(Genus, snps, FDR) %>%
  filter(FDR < 0.05) %>% 
  mutate(FDR = as.numeric(FDR)) %>%
  mutate(nlog10FDR = -log10(FDR))

Samples538_LinearAnalysisTaxaSNP_spread = Samples538_LinearAnalysisTaxaSNP_2 %>% 
  dplyr::select(Genus, snps, nlog10FDR) %>%
  tidyr::spread(key = snps, value = nlog10FDR) %>% 
  # mutate_all(as.numeric) %>%
  replace(is.na(.), 0)
Samples538_LinearAnalysisTaxaSNP_spread

Samples538_LinearAnalysisTaxaSNP_gather = Samples538_LinearAnalysisTaxaSNP_spread %>%
  gather(key = "snps", value = "nlog10FDR", -Genus) 
Samples538_LinearAnalysisTaxaSNP_gather

Samples538_LinearAnalysisTaxaSNP_gather$Genus <- abbreviate(Samples538_LinearAnalysisTaxaSNP_gather$Genus, minlength = 35, use.classes = FALSE)

# Create the heatmap
Samples538_LinearAnalysisTaxaSNP_heatmap <- ggplot(Samples538_LinearAnalysisTaxaSNP_gather, aes(x = snps, y = Genus, fill = nlog10FDR)) +
  geom_tile(color = "grey", size = 0.5) +  # Set tile borders to black with a noticeable thickness
  scale_fill_gradientn(colors = brewer.pal(9, "Reds"), name = "-log10(p-value)",
                       # limits = c(min(Samples538_LinearAnalysisTaxaSNP_gather$nlog10FDR), max(Samples538_LinearAnalysisTaxaSNP_gather$nlog10FDR)),
                       # breaks = seq(from = min(Samples538_LinearAnalysisTaxaSNP_gather$nlog10FDR), to = max(Samples538_LinearAnalysisTaxaSNP_gather$nlog10FDR), length.out = 5),
                       # labels = scales::comma
                       ) +
  theme_bw() +
  theme(
    # panel.grid.major = element_blank(),  # Remove major grid lines
    panel.grid.minor = element_blank(),  # Remove minor grid lines
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10, color = "black"),
    axis.text.y = element_text(size = 10, color = "black"),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10),
    legend.position = "bottom",  # Set legend position to top
    legend.direction = "horizontal"
  ) +
  labs( x = "Genetic Variants", y = "Microbial Species")

Samples538_LinearAnalysisTaxaSNP_heatmap
ggsave(here("Samples538_Analysis/Samples538_A06_mbQTL/Samples538_LinearAnalysisTaxaSNP_heatmap.png"),
       Samples538_LinearAnalysisTaxaSNP_heatmap,
       width = 13,
       height = 9,
       dpi = 300)

```


```{r}
histPvalueLm(Samples538_LinearAnalysisTaxaSNP)
```
```{r}
qqPlotLm(Maaslin2_16S_ITS_features_2, Samples538_recoded_vcf_2)
``` 

```{r}
Samples538_correlationMicrobes <- coringTaxa(Maaslin2_16S_ITS_features_2)
Samples538_correlationMicrobes
```


```{r}
Samples538_for_all_rsids <- allToAllProduct(Samples538_recoded_vcf_2, Maaslin2_16S_ITS_features_2)
Samples538_for_all_rsids
```


```{r}
Samples538_taxa_SNP_Cor_lim <- taxaSnpCor(Samples538_for_all_rsids,
  Samples538_correlationMicrobes,
  probs = c(0.0001, 0.9999)
)
Samples538_taxa_SNP_Cor_lim
```



```{r}
# Open a PNG device
png(here("Samples538_Analysis/Samples538_A06_mbQTL/Samples538_SNP_n_Species_Species_heatmap.png"),
         width = 3800, height = 2200, res = 300)  # Adjust size and resolution as needed

color_palette <- colorRampPalette(c("white", "yellow", "red"))(100)
mbQtlCorHeatmap(Samples538_taxa_SNP_Cor_lim,
  fontsize_col = 10,
  fontsize_row = 10,
  color = color_palette,
  angle_col = 45,
  treeheight_row = 200
)
dev.off()
```














