---
title: "Untitled"
output: html_document
date: "2023-08-10"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
This analysis will be done in another computer (BMG_SBG_308). Hence the path will be different. The final bam will be copied using
SCP transter to this computer (PC_lab).

```{bash}
cp -r /media/Second_stor/prashen/Vivianne/T2R_sequences/November_2020/raw_November_2020 /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/Raw_data/
cp -r /media/Second_stor/wasif/BuccalSwab384_samples_56_taste_genes_Nanuq_RAW_Seqs /media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/Raw_data/

```


```{bash}
# Activate conda environment
conda activate GATK_env

# make a new dir to store GATK output files
mkdir -p  /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

# Directory containing your fastq files
cd /home/mwkhan/Taste_Genes_Analysis/Raw_data/raw_November_2020


# Function to process files
process_files() {
    R1=$1
    R2=${R1/_R1.fastq.gz/_R2.fastq.gz}
    SAMPLE_NAME=$(echo $R1 | sed -e 's/NS.1414.002.TSP[0-9]*.B//' -e 's/_R1.fastq.gz//')
    OUTPUT_name=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/${SAMPLE_NAME}_FastqToSam.bam
    
    _JAVA_OPTIONS="-Xmx90g" picard FastqToSam \
        FASTQ=$R1 \
        FASTQ2=$R2 \
        OUTPUT=$OUTPUT_name \
        SAMPLE_NAME=${SAMPLE_NAME} \
        READ_GROUP_NAME=CIHR_96_9 \
        LIBRARY_NAME=Fluidigm \
        PLATFORM=illumina \
        TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}

# Export the function so it can be used by parallel
export -f process_files

# Find R1 files and pass them to parallel
ls *_R1.fastq.gz | parallel -j 10 process_files

```

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
mkdir -p Fastqc_report
mkdir -p aligned_bamqc
```


```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

process_picard_MarkIlluminaAdapters() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_FastqToSam.bam//')
  _JAVA_OPTIONS="-Xmx90g" picard MarkIlluminaAdapters \
    I=${SAMPLE_NAME}_FastqToSam.bam \
    O=${SAMPLE_NAME}_markilluminaadapters.bam \
    M=${SAMPLE_NAME}_markilluminaadapters_metrics.txt \
    TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}
# Export the function so it can be used by parallel
export -f process_picard_MarkIlluminaAdapters


#find . -name "*S1_RevertSam.bam" | parallel -j 40 process_picard_MarkIlluminaAdapters # for one file
find . -name "*_FastqToSam.bam"   | parallel -j 40 process_picard_MarkIlluminaAdapters
```


```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

# Function for processing each file
process_SamToFastq_bwa() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_markilluminaadapters.bam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  # Number of cores
  THREADS=2
  _JAVA_OPTIONS="-Xmx90g" picard SamToFastq \
      I=$1 \
      FASTQ=${SAMPLE_NAME}_samtofastq_interleaved.fq \
      CLIPPING_ATTRIBUTE=XT \
      CLIPPING_ACTION=2 \
      INTERLEAVE=true \
      NON_PF=true \
      TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
  
  /home/mwkhan/anaconda3/envs/fastqc/bin/fastqc \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq \
      -o ./Fastqc_report
  
  bwa mem -M -t $THREADS \
      -p ${REF} \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq > \
      ${SAMPLE_NAME}_bwa_mem.sam
  
  rm ${SAMPLE_NAME}_samtofastq_interleaved.fq
}

export -f process_SamToFastq_bwa

# Run the function in parallel for all *.bam files
#ls *S1_markilluminaadapters.bam  | parallel -j 10 process_SamToFastq_bwa
ls *_markilluminaadapters.bam     | parallel -j 20 process_SamToFastq_bwa
```


```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./Fastqc_report \
-o ./Fastqc_report
```

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
rm *_markilluminaadapters.bam
rm *_markilluminaadapters_metrics.txt
```

```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

# Function for processing each file
process_MergeBamAlignment() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_bwa_mem.sam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  
  _JAVA_OPTIONS="-Xmx90g" picard MergeBamAlignment \
  ALIGNED_BAM=$1 \
  UNMAPPED_BAM=${SAMPLE_NAME}_FastqToSam.bam \
  OUTPUT=${SAMPLE_NAME}_aligned.bam \
  REFERENCE_SEQUENCE=${REF} \
  CREATE_INDEX=true \
  ADD_MATE_CIGAR=true \
  CLIP_ADAPTERS=false \
  CLIP_OVERLAPPING_READS=true \
  INCLUDE_SECONDARY_ALIGNMENTS=true \
  MAX_INSERTIONS_OR_DELETIONS=-1 \
  PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
  ATTRIBUTES_TO_RETAIN=XS \
  TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}

export -f process_MergeBamAlignment

ls *_bwa_mem.sam     | parallel -j 10 process_MergeBamAlignment
```


```{bash}
mkdir /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_bedfiles
# copy appropriate bed fileto the working directory
# New_hg38_Project_373.bed
copy /media/Second_stor/prashen/Vivianne/T2R_sequences/November_2020/Bed153Fluidigm.bed \
/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_bedfiles/CIHR_96_9_genes_153.bed
```


```{bash}
<!-- conda activate qualimap_env -->

<!-- cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK -->

<!-- rm -r    ./aligned_bamqc/Qualimap -->
<!-- mkdir -p ./aligned_bamqc/Qualimap -->

<!-- process_Qualimap() { -->
<!--   SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//') -->

<!--     _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \ -->
<!--       -nt 1 \ -->
<!--       -bam ${SAMPLE_NAME}_aligned.bam \ -->
<!--       -outdir ./aligned_bamqc/Qualimap/${SAMPLE_NAME}_aligned_qualimap \ -->
<!--       --genome-gc-distr hg19 \ -->
<!--       --feature-file /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_bedfiles/CIHR_96_9_genes_153_hg38.bed -->
<!-- } -->

<!-- export -f process_Qualimap -->
<!-- # Run the function in parallel for all *.bam files -->
<!-- ls *_aligned.bam     | parallel -j 5 process_Qualimap -->



<!-- cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK -->

<!-- /home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \ -->
<!-- --interactive --force \ -->
<!-- ./aligned_bamqc/Qualimap \ -->
<!-- -o ./aligned_bamqc/Qualimap -->
```


```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

# Function for processing each file
process_picard_SortSam() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')
  
  _JAVA_OPTIONS="-Xmx80g" picard SortSam \
      I=${SAMPLE_NAME}_aligned.bam \
      O=${SAMPLE_NAME}_sorted.bam \
      SORT_ORDER=coordinate \
  TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK
}
export -f process_picard_SortSam

# Run the function in parallel for all *.bam files
ls *_aligned.bam  | parallel -j 10 process_picard_SortSam
```


## sorted bam files are copied to VERBATIM_HD_PC as back up.
## Use these sam files from VERBATIM_HD_PC to generate sorted_BQSR files in the computer 

```{bash}
# Activate conda environment
conda activate GATK_env
cd /media/wasif_pclab/VERBATIM_HD_PC/Samples550_Analysis/Samples550_sorted_sam_backup/Samples96_genes9_sorted_sam

out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR
rm -r     ${out_dir}
mkdir -p  ${out_dir}

# Function for processing each file
process_gatk_BaseRecalibrator() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  
  out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR
  
  ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
  
  _JAVA_OPTIONS="-Xmx90g" gatk BaseRecalibrator \
      -R $ref_file \
      -I $1 \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table

  _JAVA_OPTIONS="-Xmx90g" gatk ApplyBQSR \
      -R $ref_file \
      -I $1 \
      --bqsr-recal-file ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_BQSR.bam
}

export -f process_gatk_BaseRecalibrator
ls   *_sorted.bam   | parallel -j 20 process_gatk_BaseRecalibrator
```

```{bash}
conda activate qualimap_env

unset DISPLAY
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR

rm -r    ./sorted_BQSR_bamqc/Qualimap
mkdir -p ./sorted_BQSR_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_BQSR.bam \
      -outdir ./sorted_BQSR_bamqc/Qualimap/${SAMPLE_NAME}_sorted_BQSR_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_bedfiles/CIHR_96_9_genes_153_hg38.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_BQSR.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/wasif_pclab/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_BQSR_bamqc/Qualimap \
-o ./sorted_BQSR_bamqc/Qualimap
```

## to save space, these files are backed up in VERBATIM_HD_PC

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
  rm *_sorted.bam_sorted_BQSR.bam
  rm *_sorted.bam_sorted_BQSR.bai

```

## The previous analysis did not include deduplication using markduplicates option
below will be the analysis based on markduplicates. For that I am planning to first merge
sorted.bam files for different runs for the same samples
to do that the bam files should have proper gatkdictionary i.e same name in @RG header in the bam files
This analysis will be done in Jin_SBG_BMG computer

```{bash}
# Activate conda environment
conda activate GATK_env

cd /home/mwkhan/Taste_Genes_Analysis/Samples550_sorted_sam_backup/Samples96_genes9_sorted_sam


# Function for processing each file
process_picard_AddOrReplaceReadGroups() {

  work_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
  mkdir -p /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  SAMPLE_number=$(echo $1 | sed -n 's/^S\([0-9]*\).*$/\1/p')
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" picard AddOrReplaceReadGroups \
        I=$1 \
        O=${work_dir}/${SAMPLE_NAME}_sorted_RG.bam \
        RGSM=CIHR_$SAMPLE_number \
        RGLB=CIHR_96_9 \
        RGPL=illumina \
        RGPU=CIHR_96_9
}

export -f process_picard_AddOrReplaceReadGroups
ls   *_sorted.bam  | parallel -j 40 process_picard_AddOrReplaceReadGroups
```





