---
title: "Untitled"
output: html_document
date: "2023-08-10"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
This analysis will be done in another computer (BMG_SBG_308). Hence the path will be different. The final bam will be copied using
SCP transter to this computer (PC_lab).

```{bash}

# to change CHRIM_80 to CIHR_96
# 22 to 46
```


```{bash}
# Activate conda environment
conda activate GATK_env

# make a new dir to store GATK output files
mkdir -p  /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/


process_picard_RevertSam() {
  SAMPLE_NAME=$(basename "$1" | sed -E 's/.*\CIHR-(S[0-9]+)\.bam/\1/')
# RevertSam process to be given to parallel processing

out_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  _JAVA_OPTIONS="-Xmx90g" picard RevertSam \
    I=$1 \
    O=${out_dir}/${SAMPLE_NAME}_RevertSam.bam \
    SANITIZE=true \
    MAX_DISCARD_FRACTION=0.005 \
    ATTRIBUTE_TO_CLEAR=XT \
    ATTRIBUTE_TO_CLEAR=XN \
    ATTRIBUTE_TO_CLEAR=AS \
    ATTRIBUTE_TO_CLEAR=OC \
    ATTRIBUTE_TO_CLEAR=OP \
    ATTRIBUTE_TO_CLEAR=XS \
    SORT_ORDER=queryname  \
    RESTORE_ORIGINAL_QUALITIES=true \
    REMOVE_DUPLICATE_INFORMATION=true \
    REMOVE_ALIGNMENT_INFORMATION=true
}
# Export the function so it can be used by parallel
export -f process_picard_RevertSam

# pass them to parallel
in_dir=/home/mwkhan/Taste_Genes_Analysis/Raw_data/Fluidigm_373_CIHR_samples_96
# try with one files
#ls $in_dir/*S1.bam   | parallel -j 20 process_picard_RevertSam
ls $in_dir/*.bam     | parallel -j 10 process_picard_RevertSam

# takes about 30 minutes


```

## add the new folder to the .gitignore file to avoid git uploading of all the gatk intermediate files

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
mkdir -p Fastqc_report
mkdir -p aligned_bamqc
```


```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

process_picard_MarkIlluminaAdapters() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_RevertSam.bam//')
  _JAVA_OPTIONS="-Xmx90g" picard MarkIlluminaAdapters \
    I=${SAMPLE_NAME}_RevertSam.bam \
    O=${SAMPLE_NAME}_markilluminaadapters.bam \
    M=${SAMPLE_NAME}_markilluminaadapters_metrics.txt
}
# Export the function so it can be used by parallel
export -f process_picard_MarkIlluminaAdapters


#find . -name "*S1_RevertSam.bam" | parallel -j 40 process_picard_MarkIlluminaAdapters # for one file
find . -name "*_RevertSam.bam"   | parallel -j 40 process_picard_MarkIlluminaAdapters
```

## copy reference folder

```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

# Function for processing each file
process_SamToFastq_bwa() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_markilluminaadapters.bam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  # Number of cores
  THREADS=2
  _JAVA_OPTIONS="-Xmx90g" picard SamToFastq \
      I=$1 \
      FASTQ=${SAMPLE_NAME}_samtofastq_interleaved.fq \
      CLIPPING_ATTRIBUTE=XT \
      CLIPPING_ACTION=2 \
      INTERLEAVE=true \
      NON_PF=true
  
  /home/mwkhan/anaconda3/envs/fastqc/bin/fastqc \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq \
      -o ./Fastqc_report
  
  bwa mem -M -t $THREADS \
      -p ${REF} \
      ${SAMPLE_NAME}_samtofastq_interleaved.fq > \
      ${SAMPLE_NAME}_bwa_mem.sam
  
  rm ${SAMPLE_NAME}_samtofastq_interleaved.fq
}

export -f process_SamToFastq_bwa

# Run the function in parallel for all *.bam files
#ls *S1_markilluminaadapters.bam  | parallel -j 10 process_SamToFastq_bwa
ls *_markilluminaadapters.bam     | parallel -j 20 process_SamToFastq_bwa
```

```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./Fastqc_report \
-o ./Fastqc_report
```


```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
rm *_markilluminaadapters.bam
rm *_markilluminaadapters_metrics.txt
```


```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

# Function for processing each file
process_MergeBamAlignment() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_bwa_mem.sam//')
  # Path to your reference genome
  REF="/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna"

  
  _JAVA_OPTIONS="-Xmx90g" picard MergeBamAlignment \
  ALIGNED_BAM=$1 \
  UNMAPPED_BAM=${SAMPLE_NAME}_RevertSam.bam \
  OUTPUT=${SAMPLE_NAME}_aligned.bam \
  REFERENCE_SEQUENCE=${REF} \
  CREATE_INDEX=true \
  ADD_MATE_CIGAR=true \
  CLIP_ADAPTERS=false \
  CLIP_OVERLAPPING_READS=true \
  INCLUDE_SECONDARY_ALIGNMENTS=true \
  MAX_INSERTIONS_OR_DELETIONS=-1 \
  PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
  ATTRIBUTES_TO_RETAIN=XS
}

export -f process_MergeBamAlignment

# Run the function in parallel for all *.bam files
#ls *S1_markilluminaadapters.bam  | parallel -j 10 process_MergeBamAlignment
ls *_bwa_mem.sam     | parallel -j 5 process_MergeBamAlignment
```

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  rm *_bwa_mem.sam
  rm *_RevertSam.bam
```


```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

rm ./Fastqc_report/*_samtofastq_interleaved_fastqc.zip
rm ./Fastqc_report/*_samtofastq_interleaved_fastqc.html
```

```{bash}
mkdir /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_bedfiles
# copy appropriate bed fileto the working directory
# New_hg38_Project_373.bed
<!-- copy /media/Second_stor/prashen/Vivianne/T2R_sequences/December_2021/New_hg38_Project_293.bed \ -->
<!-- /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_bedfiles/CIHR_96_46_genes_293.bed -->
```

```{bash}
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  rm *_markilluminaadapters.bam
  rm *_markilluminaadapters_metrics.txt
```

```{bash}
conda activate qualimap_env

cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

rm -r    ./Aligned_bamqc/Qualimap
mkdir -p ./Aligned_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')
  
    _JAVA_OPTIONS="-Xmx90g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_aligned.bam \
      -outdir ./Aligned_bamqc/Qualimap/${SAMPLE_NAME}_aligned_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_bedfiles/New_hg38_Project_373_no_names.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_aligned.bam     | parallel -j 5 process_Qualimap

```

## MultiQC
```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./Aligned_bamqc/Qualimap \
-o ./Aligned_bamqc/Qualimap

```


```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

# Function for processing each file
process_picard_SortSam() {
  file=$1
  SAMPLE_NAME=$(echo $1 | sed -e 's/_aligned.bam//')
  
  _JAVA_OPTIONS="-Xmx80g" picard SortSam \
      I=${SAMPLE_NAME}_aligned.bam \
      O=${SAMPLE_NAME}_sorted.bam \
      SORT_ORDER=coordinate
}
export -f process_picard_SortSam

# Run the function in parallel for all *.bam files
ls *_aligned.bam  | parallel -j 10 process_picard_SortSam
```

## if everything looks good then previous files can be deleted
```{bash}
 cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  rm *_aligned.bam
  rm *_aligned.bai
```

## sorted bam files are copied to VERBATIM_HD_PC as back up.
## Use these sam files from VERBATIM_HD_PC to generate sorted_BQSR files in the computer 

```{bash}
# Activate conda environment
conda activate GATK_env
cd /media/wasif_pclab/VERBATIM_HD_PC/Samples550_Analysis/Samples550_sorted_sam_backup/Samples96_genes46_sorted_sam

out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
rm -r     ${out_dir}
mkdir -p  ${out_dir}

# Function for processing each file
process_gatk_BaseRecalibrator() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  
  out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
  
  ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
  
  _JAVA_OPTIONS="-Xmx90g" gatk BaseRecalibrator \
      -R $ref_file \
      -I $1 \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf \
      --known-sites /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table

  _JAVA_OPTIONS="-Xmx90g" gatk ApplyBQSR \
      -R $ref_file \
      -I $1 \
      --bqsr-recal-file ${out_dir}/${SAMPLE_NAME}_sorted_recal_data.table \
      -O ${out_dir}/${SAMPLE_NAME}_sorted_BQSR.bam
}

export -f process_gatk_BaseRecalibrator
#ls   *_sorted.bam   | parallel -j 20 process_gatk_BaseRecalibrator
ls   S3_sorted.bam S4_sorted.bam  | parallel -j 20 process_gatk_BaseRecalibrator
```

```{bash}
conda activate qualimap_env

unset DISPLAY
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR

rm -r    ./sorted_BQSR_bamqc/Qualimap
mkdir -p ./sorted_BQSR_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_BQSR.bam \
      -outdir ./sorted_BQSR_bamqc/Qualimap/${SAMPLE_NAME}_sorted_BQSR_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_bedfiles/New_hg38_Project_373.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_BQSR.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/wasif_pclab/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_BQSR_bamqc/Qualimap \
-o ./sorted_BQSR_bamqc/Qualimap
```

## to save space, these files are backed up in VERBATIM_HD_PC

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  rm *_sorted.bam_sorted_BQSR.bam
  rm *_sorted.bam_sorted_BQSR.bai
```


## The previous analysis did not include deduplication using markduplicates option
below will be the analysis based on markduplicates. Since we need to merge
sorted.bam files for different runs for the same samples later,
to do that the bam files should have proper gatkdictionary i.e same name in @RG header in the bam files
This analysis will be done in Jin_SBG_BMG computer

```{bash}
# Activate conda environment
conda activate GATK_env

cd /home/mwkhan/Taste_Genes_Analysis/Samples550_sorted_sam_backup/Samples96_genes46_sorted_sam


# Function for processing each file
process_picard_AddOrReplaceReadGroups() {

  work_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
  mkdir -p /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted.bam//')
  SAMPLE_number=$(echo $1 | sed -n 's/^S\([0-9]*\).*$/\1/p')
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" picard AddOrReplaceReadGroups \
        I=$1 \
        O=${work_dir}/${SAMPLE_NAME}_sorted_RG.bam \
        RGSM=CIHR_$SAMPLE_number \
        RGLB=CIHR_96_46 \
        RGPL=illumina \
        RGPU=CIHR_96_46
}

export -f process_picard_AddOrReplaceReadGroups
ls   *_sorted.bam  | parallel -j 40 process_picard_AddOrReplaceReadGroups
```

## Mark duplicates
```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK

process_MarkDuplicates() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_RG.bam//')

  # _JAVA_OPTIONS="-Xmx90g" picard MarkDuplicates \
  #     INPUT=$1 \
  #     OUTPUT=${SAMPLE_NAME}_sorted_RG_dedup.bam \
  #     M=${SAMPLE_NAME}_sorted_RG_dedup.txt \
  #     TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK/
      
  samtools rmdup -s \
  $1 \
  ${SAMPLE_NAME}_sorted_RG_rmdup.bam \
  --verbosity 2
}
export -f process_MarkDuplicates
ls   *_sorted_RG.bam   | parallel -j 40 process_MarkDuplicates
# for picard MarkDuplicates, do not use more than 2 threads
# for samtools rmdup, 40 threads can be used
```