---
title: "538 vcf annotation, quality, and visualisation"
output: html_document
date: "2024-03-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{bash, eval=FALSE}
#bcftools annotate requires bgzipped files
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK

bgzip \
-c ./Samples538_55genes_GenotypeGVCFs.vcf > \
./Samples538_55genes_GenotypeGVCFs.vcf.gz

tabix \
-p vcf \
Samples538_55genes_GenotypeGVCFs.vcf.gz

bcftools annotate \
-a /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf.gz \
-c ID \
Samples538_55genes_GenotypeGVCFs.vcf.gz > \
Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf
bcftools view -H Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf | wc -l
```


```{bash,eval=FALSE}
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK/Samples538_VCFqual

## Calculate average sampling depth per samples: Use file before annotation and filter because the filtered file doesn't have the SD info:

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK/Samples538_VCFqual
file_prefix=Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf


# Calculate allele frequency
vcftools --vcf ../$file_prefix --out $file_prefix --freq2  --max-alleles 2
# Calculate mean depth per individual
vcftools --vcf ../$file_prefix --out $file_prefix --depth
# Calculate mean depth per site
vcftools --vcf ../$file_prefix --out $file_prefix --site-mean-depth
# Calculate site quality
vcftools --vcf ../$file_prefix --out $file_prefix --site-quality 
# Calculate proportion of missing data per individual
vcftools --vcf ../$file_prefix --out $file_prefix --missing-indv
# Calculate proportion of missing data per site
vcftools --vcf ../$file_prefix --out $file_prefix --missing-site
# Calculate heterozygosity and inbreeding coefficient per individual
vcftools --vcf ../$file_prefix --out $file_prefix --het 

```

# Variant quality
here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf
```{r}
var_qual <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.lqual"),
                       delim = "\t", col_names = c("chr", "pos", "qual"), skip = 1)
a <- ggplot(var_qual, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 100)
```

# Variant mean depth
```{r}
var_depth <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.ldepth.mean"),
                        delim = "\t", col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)
a <- ggplot(var_depth, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

summary(var_depth$mean_depth)

a + theme_light() + xlim(0, 400)
```

# Variant missingness
```{r}
var_miss <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.lmiss"),
                       delim = "\t", col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(var_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 0.05)

summary(var_miss$fmiss)
```

# Minor allele frequency
```{r}
var_freq <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.frq"),
                       delim = "\t", col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)
# find minor allele frequency
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))
a <- ggplot(var_freq, aes(maf)) + geom_bar(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 0.6) + ylim(0, 10)
summary(var_freq$maf)
```

# Individual based statistics
# Mean depth per individual
```{r}
ind_depth <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.idepth"),
                        delim = "\t", col_names = c("ind", "nsites", "depth"), skip = 1)
a <- ggplot(ind_depth, aes(depth)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```

# Proportion of missing data per individual
```{r}
ind_miss  <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.imiss"),
                        delim = "\t", col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(ind_miss, aes(fmiss)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```

# Heterozygosity and inbreeding coefficient per individual
```{r}
ind_het <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs.vcf.het"),
                      delim = "\t", col_names = c("ind","ho", "he", "nsites", "f"), skip = 1)
a <- ggplot(ind_het, aes(f)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```


# Applying filters to the VCF
```{bash, eval=FALSE}
<!-- vcftools defults -->
<!--          --maf 0 -->
<!--         --maxDP 150 -->
<!--         --max-meanDP 150 -->
<!--         --minDP 5 -->
<!--         --min-meanDP 5 -->
<!--         --minQ 30 -->
<!--         --max-missing 0.5 -->
        
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK
root_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK
VCF_IN=$root_dir/Samples538_55genes_GenotypeGVCFs_SNPIDs_anno.vcf
VCF_OUT=$root_dir/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf

# set filters
MAF=0.00
MAX_MISS=0.5 # default 0.5, somewhat counter-intuitive 0 is totally missing, 1 is none missing
QUAL=25 # default 30
MIN_DEPTH=5 # default 5
MAX_DEPTH=150 # # default 150


# perform the filtering with vcftools
vcftools \
--vcf $VCF_IN \
--maf $MAF \
--max-missing $MAX_MISS \
--minQ $QUAL \
--min-meanDP $MIN_DEPTH \
--minDP $MIN_DEPTH \
--recode \
--stdout > $VCF_OUT
# other options
#--max-meanDP $MAX_DEPTH \
#--maxDP $MAX_DEPTH \
# --remove-indels \
cat out.log
echo "number of SNPs after the above filtering: "
bcftools view -H $VCF_OUT | wc -l
```

```{bash, eval=FALSE}
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK/Samples538_VCFqual

## Calculate average sampling depth per samples: Use file before annotation and filter because the filtered file doesn't have the SD info:

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK/Samples538_VCFqual
file_prefix=Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf


# Calculate allele frequency
vcftools --vcf ../$file_prefix --out $file_prefix --freq2  --max-alleles 2
# Calculate mean depth per individual
vcftools --vcf ../$file_prefix --out $file_prefix --depth
# Calculate mean depth per site
vcftools --vcf ../$file_prefix --out $file_prefix --site-mean-depth
# Calculate site quality
vcftools --vcf ../$file_prefix --out $file_prefix --site-quality 
# Calculate proportion of missing data per individual
vcftools --vcf ../$file_prefix --out $file_prefix --missing-indv
# Calculate proportion of missing data per site
vcftools --vcf ../$file_prefix --out $file_prefix --missing-site
# Calculate heterozygosity and inbreeding coefficient per individual
vcftools --vcf ../$file_prefix --out $file_prefix --het 

```

# Variant quality
here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf
```{r}
var_qual <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.lqual"),
                       delim = "\t", col_names = c("chr", "pos", "qual"), skip = 1)
a <- ggplot(var_qual, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 100)
```

# Variant mean depth
```{r}
var_depth <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.ldepth.mean"),
                        delim = "\t", col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)
a <- ggplot(var_depth, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

summary(var_depth$mean_depth)

a + theme_light() + xlim(0, 400)
```

# Variant missingness
```{r}
var_miss <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.lmiss"),
                       delim = "\t", col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(var_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 0.05)

summary(var_miss$fmiss)
```

# Minor allele frequency
```{r}
var_freq <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.frq"),
                       delim = "\t", col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)
# find minor allele frequency
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))
a <- ggplot(var_freq, aes(maf)) + geom_bar(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 0.6) + ylim(0, 10)
summary(var_freq$maf)
```

# Individual based statistics
# Mean depth per individual
```{r}
ind_depth <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.idepth"),
                        delim = "\t", col_names = c("ind", "nsites", "depth"), skip = 1)
a <- ggplot(ind_depth, aes(depth)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```

# Proportion of missing data per individual
```{r}
ind_miss  <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.imiss"),
                        delim = "\t", col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(ind_miss, aes(fmiss)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```

# Heterozygosity and inbreeding coefficient per individual
```{r}
ind_het <- read_delim(here("Samples538_Analysis/Samples538_GATK/Samples538_VCFqual/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf.het"),
                      delim = "\t", col_names = c("ind","ho", "he", "nsites", "f"), skip = 1)
a <- ggplot(ind_het, aes(f)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
```


# annotate the VCF with gene names
```{bash, eval=FALSE}
conda activate SnpEff_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK

snpEff \
-v hg38 \
./Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered.vcf> \
./Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf
```

# fill empty SNP IDs with chr and BP
When we read a vcf using VariantAnnotation::readVcf it automatically fills the SNP IDs with chr and BP. So that file can be directly saved as a vcf file.
This not working as it removes the * given in the ALT column
```{r}
# vcf <- VariantAnnotation::readVcf(here("Samples538_Analysis/Samples538_GATK/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf"))
# # some IDs are chr1:1285974_C/*	C	,A which gives and error due to * in the ID
# 
# # to replace them
# rownames(vcf) <- gsub(pattern = "\\*", replacement = "", x = rownames(vcf))
# VariantAnnotation::writeVcf(vcf, file = here("Samples538_Analysis/Samples538_GATK/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno_IDfilled.vcf"))

```
```{bash, eval=FALSE}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK
awk 'BEGIN{FS=OFS="\t"} !/^#/ {if($3 == ".") $3 = $1"_"$2} {print}' Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf \
> Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno_IDcomplete.vcf
```

```{bash, eval=FALSE}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK
awk 'BEGIN{FS=OFS="\t"} !/^#/ {$3 = $1"_"$2"_"$3} {print}' Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf \
> Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno_IDcomplete2.vcf
```


```{bash, eval=FALSE}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples538_Analysis/Samples538_GATK
awk \
'BEGIN {FS=OFS="\t"} !/^#/ {split($8, parts, "|"); gene_info = (length(parts) >= 5 ? parts[5] : "unknown"); $3 = $1"_"$2"_"gene_info"_"$3; print;} /^#/ {print}' \
Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf > \
Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno_IDcomplete2.vcf

```


```{r}
# Load the VariantAnnotation package
library(VariantAnnotation)

# Specify the path to your VCF file
vcf_file <- here("Samples538_Analysis/Samples538_GATK/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf")

# Read the VCF file
vcf <- readVcf(vcf_file)

# Extract gene names
# This assumes gene names are stored in the INFO field under 'GENEINFO'
info(vcf)
```

```{r}
library(vcfR)
my_vcf <- read.vcfR(here("Samples538_Analysis/Samples538_GATK/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf"))
my_vcf

```

```{r}
# Load the VariantAnnotation package
library(VariantAnnotation)

# Specify the path to your VCF file
vcf_file <- here("Samples538_Analysis/Samples538_GATK/Samples538_55genes_GenotypeGVCFs_SNPIDs_filtered_anno.vcf")

# Read the VCF file
vcf <- readVcf(vcf_file)

# Assuming gene info is in 'GENEINFO' field; adjust as necessary for your VCF annotations
gene_info <- info(vcf)$GENEINFO

# Extract chromosomes, positions, and original IDs
chromosomes <- seqnames(vcf)
positions <- start(vcf)
original_ids <- rownames(vcf)

# Create new IDs by combining the chromosome, position, gene info, and original ID
# Adjust the format as needed
new_ids <- paste(chromosomes, positions, gene_info, original_ids, sep="_")

# Assign new IDs back to the VCF object
rownames(vcf) <- new_ids

# If you want to write this modified VCF to a file
output_file <- "modified_vcf_file.vcf"
writeVcf(vcf, output_file)
```

