---
title: "Untitled"
output: html_document
date: "2023-08-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
```{bash}

conda activate GATK_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis
work_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK
mkdir -p $work_dir
cd $work_dir

process_picard_MergeSamFiles() {
CIHR_96_46=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
CIHR_96_9=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR

 SAMPLE_NAME=$(basename "$1" | sed -e 's/_sorted_BQSR.bam//')
 
 _JAVA_OPTIONS="-Xmx90g" picard MergeSamFiles \
      I=${CIHR_96_46}/${SAMPLE_NAME}_sorted_BQSR.bam \
      I=${CIHR_96_9}/${SAMPLE_NAME}_sorted_BQSR.bam \
      O=${SAMPLE_NAME}_sorted_BQSR_55genes.bam 
}

# Export the function so it can be used by parallel
export -f process_picard_MergeSamFiles

# pass them to parallel
in_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
# try with one files
ls $in_dir/*_sorted_BQSR.bam   | parallel -j 40 process_picard_MergeSamFiles

```

2 files do not have 46 genes and 2 files for 9 genes is missing.
copy them with changed name as in MergeSamFiles so that they can be processed together in the downstream analysis
```{bash}
cp_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
cp ./S3_sorted_BQSR.bam ${cp_dir}/S3_sorted_BQSR_55genes.bam
cp ./S4_sorted_BQSR.bam ${cp_dir}/S4_sorted_BQSR_55genes.bam

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR
cp ./S2_sorted_BQSR.bam ${cp_dir}/S2_sorted_BQSR_55genes.bam
cp ./S9_sorted_BQSR.bam ${cp_dir}/S9_sorted_BQSR_55genes.bam
```


```{bash}
conda activate qualimap_env

unset DISPLAY
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK

rm -r    ./sorted_BQSR_55genes_bamqc/Qualimap
mkdir -p ./sorted_BQSR_55genes_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR_55genes.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_BQSR_55genes.bam \
      -outdir ./sorted_BQSR_55genes_bamqc/Qualimap/${SAMPLE_NAME}_sorted_BQSR_55genes_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/bed_files/Merged_bed_hg38_Jan15.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_BQSR_55genes.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/wasif_pclab/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_BQSR_55genes_bamqc/Qualimap \
-o ./sorted_BQSR_55genes_bamqc/Qualimap
```

## remove the bam files not required to save space 
```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK/sorted_BQSR
rm *_sorted_BQSR.bam
rm *_sorted_recal_data.table
rm *_sorted_BQSR.bai
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK/sorted_BQSR
rm *_sorted_BQSR.bam
rm *_sorted_recal_data.table
rm *_sorted_BQSR.bai
```

## After merging bam files, they have two headers 
e.g. 
samtools view -H  /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/S1_sorted_BQSR_55genes.bam

@RG     ID:HSP_VAL12407421-WS-A01_1148_2        SM:S1   LB:HSP_VAL12407421-WS-A01       PU:run1148_2    CN:McGill University and Genome Quebec Innovation Centre        PL:Illumina
@RG     ID:HSP_VAL12506519-WS-A-A01_1713_2      SM:CIHR_S1     LB:HSP_VAL12506519-WS-A-A01     PU:run1713_2    CN:Centre expertise et de services Genome Quebec        PL:Illumina
has these two @RG header with different sample names.
# to fix this

<!-- ##try delete  -->
<!-- ```{bash} -->
<!-- _JAVA_OPTIONS="-Xmx90g" picard MergeSamFiles \ -->
<!--       I=/media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/Samples550_sorted_sam_backup/Samples80_genes35_sorted_sam/S1_sorted.bam \ -->
<!--       I=/media/wasif_pclab/VERBATIM_HD_PC/Samples384_Analysis/Samples550_sorted_sam_backup/Samples80_genes22_sorted_sam/S1_sorted.bam \ -->
<!--       O=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/try_delete/S1_sorted_BQSR_55genes_try_delete.bam \ -->
<!--       MERGE_SEQUENCE_DICTIONARIES=true -->

<!-- ## check if there is one or two @RG headers -->
<!-- samtools view -H /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/try_delete/S1_sorted_BQSR_55genes_try_delete.bam -->

<!-- # We still have two different @RG headers in the above bam file so it is better to -->
<!-- # use the AddOrReplaceReadGroups -->
<!-- ``` -->


```{bash }
# Activate conda environment
conda activate GATK_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK

# Function for processing each file
process_picard_AddOrReplaceReadGroups() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR_55genes.bam//')
  SAMPLE_number=$(echo $1 | sed -n 's/^S\([0-9]*\).*$/\1/p')
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" picard AddOrReplaceReadGroups \
        I=$1 \
        O=${SAMPLE_NAME}_sorted_BQSR_RG_55genes.bam \
        RGSM=CIHR_$SAMPLE_number \
        RGLB=lib1 \
        RGPL=illumina \
        RGPU=unit1
}

export -f process_picard_AddOrReplaceReadGroups
#ls    *_sorted_BQSR_55genes.bam  | parallel -j 40 process_picard_AddOrReplaceReadGroups
ls    S2_sorted_BQSR_55genes.bam S3_sorted_BQSR_55genes.bam S4_sorted_BQSR_55genes.bam S9_sorted_BQSR_55genes.bam | parallel -j 40 process_picard_AddOrReplaceReadGroups
```


```{bash}
# Activate conda environment
conda activate GATK_env

cd        /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/sorted_BQSR_RG_55genes_HaplotypeCaller

# Function for processing each file
process_gatk_HaplotypeCaller() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_BQSR_RG_55genes.bam//')
  
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" gatk HaplotypeCaller \
      -R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
      -I $1 \
      -O sorted_BQSR_RG_55genes_HaplotypeCaller/${SAMPLE_NAME}_raw_variants.g.vcf.gz \
      --assembly-region-padding 100 \
      -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/bed_files/Merged_bed_hg38_Jan15_no_names.bed \
      -ERC GVCF
}

export -f process_gatk_HaplotypeCaller
#ls    *_sorted_BQSR_RG_55genes.bam  | parallel -j 40 process_gatk_HaplotypeCaller
ls    S2_sorted_BQSR_RG_55genes.bam S3_sorted_BQSR_RG_55genes.bam S4_sorted_BQSR_RG_55genes.bam S9_sorted_BQSR_RG_55genes.bam | parallel -j 40 process_gatk_HaplotypeCaller
```

## copy to external hard drive to save space
```{bash}
find /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/ \
-type f \( -name "*.bam" -o -name "*.bai" \) | parallel -j 40 \
--no-notice cp {} /media/wasif_pclab/VERBATIM_HD_PC/Samples550_Analysis/Sample550_sorted_BQSR_RG_backup/Sample96_55genes_sorted_BQSR_RG_backup/

# first make sure the files are properly copied
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/
rm ./*.{bam,bai}

```

## This is for dedup analysis
# merge the two runs

```{bash}

conda activate Picard_env

work_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK
mkdir -p $work_dir
cd $work_dir

process_picard_MergeSamFiles() {
CIHR_96_46=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
CIHR_96_9=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK

 SAMPLE_NAME=$(basename "$1" | sed -e 's/_sorted_RG.bam//')
 
 _JAVA_OPTIONS="-Xmx90g" picard MergeSamFiles \
      I=${CIHR_96_46}/${SAMPLE_NAME}_sorted_RG.bam \
      I=${CIHR_96_9}/${SAMPLE_NAME}_sorted_RG.bam \
      O=${SAMPLE_NAME}_sorted_RG_55genes.bam 
}

# Export the function so it can be used by parallel
export -f process_picard_MergeSamFiles

# pass them to parallel
in_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
in_dir2=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
# try with one files
ls $in_dir/*_sorted_RG.bam  | parallel -j 40 process_picard_MergeSamFiles

```

2 files do not have 46 genes and 2 files for 9 genes is missing.
copy them with changed name as in MergeSamFiles so that they can be processed together in the downstream analysis
```{bash}
cp_dir=/home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK

cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_46/CIHR_96_46_GATK
cp ./S3_sorted_RG.bam ${cp_dir}/S3_sorted_RG_55genes.bam
cp ./S4_sorted_RG.bam ${cp_dir}/S4_sorted_RG_55genes.bam

cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_9/CIHR_96_9_GATK
cp ./S2_sorted_RG.bam ${cp_dir}/S2_sorted_RG_55genes.bam
cp ./S9_sorted_RG.bam ${cp_dir}/S9_sorted_RG_55genes.bam
```


## Mark duplicates
```{bash}
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK

process_MarkDuplicates() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_RG_55genes.bam//')

  # _JAVA_OPTIONS="-Xmx90g" picard MarkDuplicates \
  #     INPUT=$1 \
  #     OUTPUT=${SAMPLE_NAME}_sorted_RG_dedup.bam \
  #     M=${SAMPLE_NAME}_sorted_RG_dedup.txt \
  #     TMP_DIR=/home/mwkhan/Taste_Genes_Analysis/tmp_dir_for_GATK/
      
  # samtools rmdup -s \
  # $1 \
  # ${SAMPLE_NAME}_sorted_RG_rmdup_55genes.bam \
  # --verbosity 2
  
  # rmdup did not remove any reads so trying markdup

    samtools sort -n $1 | \
    samtools fixmate -m - - | \
    samtools sort - | \
    samtools markdup -r - ${SAMPLE_NAME}_sorted_RG_55genes_markdup.bam \
    --verbosity 2
  
}
export -f process_MarkDuplicates
ls   *_sorted_RG_55genes.bam   | parallel -j 40 process_MarkDuplicates
# for picard MarkDuplicates, do not use more than 2 threads
# for samtools rmdup, 40 threads can be used
# for samtools markdup, 40 threads can be used takes about 1 hour
```

```{bash}
conda activate qualimap_env

unset DISPLAY
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/

rm -r    ./sorted_RG_55genes_markdup_bamqc/Qualimap
mkdir -p ./sorted_RG_55genes_markdup_bamqc/Qualimap

process_Qualimap() {
  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_RG_55genes_markdup.bam//')
  
    _JAVA_OPTIONS="-Xms10g -Xmx30g" qualimap bamqc \
      -nt 1 \
      -bam ${SAMPLE_NAME}_sorted_RG_55genes_markdup.bam \
      -outdir ./sorted_RG_55genes_markdup_bamqc/Qualimap/${SAMPLE_NAME}_sorted_RG_55genes_markdup_qualimap \
      --genome-gc-distr hg19 \
      --feature-file /home/mwkhan/Taste_Genes_Analysis/References_Seqs/bed_files/Merged_bed_hg38_Jan15_no_names.bed
}

export -f process_Qualimap
# Run the function in parallel for all *.bam files
ls *_sorted_RG_55genes_markdup.bam     | parallel -j 8 process_Qualimap

## MultiQC
/home/mwkhan/anaconda3/envs/Multiqc_env/bin/multiqc \
--interactive --force \
./sorted_RG_55genes_markdup_bamqc/Qualimap \
-o ./sorted_RG_55genes_markdup_bamqc/Qualimap
```

```{bash}
# Activate conda environment
conda activate GATK_env
cd /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK



# Function for processing each file
process_gatk_BaseRecalibrator() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_RG_55genes_markdup.bam//')
  

  ref_file=/home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
  
  _JAVA_OPTIONS="-Xmx90g" gatk BaseRecalibrator \
      -R $ref_file \
      -I $1 \
      --known-sites /home/mwkhan/Taste_Genes_Analysis/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf \
      --known-sites /home/mwkhan/Taste_Genes_Analysis/References_Seqs/Homo_sapiens_assembly38.known_indels.vcf \
      -O ${SAMPLE_NAME}_sorted_recal_data.table

  _JAVA_OPTIONS="-Xmx90g" gatk ApplyBQSR \
      -R $ref_file \
      -I $1 \
      --bqsr-recal-file ${SAMPLE_NAME}_sorted_recal_data.table \
      -O ${SAMPLE_NAME}_sorted_RG_55genes_markdup_BQSR.bam
}

export -f process_gatk_BaseRecalibrator
ls   *_sorted_RG_55genes_markdup.bam   | parallel -j 20 process_gatk_BaseRecalibrator
```




```{bash}
# Activate conda environment
conda activate GATK_env

cd        /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/
mkdir -p  /home/mwkhan/Taste_Genes_Analysis/CIHR_96_Analysis/CIHR_96_genes_55/CIHR_96_55_GATK/sorted_RG_55genes_markdup_BQSR_HaplotypeCaller

# Function for processing each file
process_gatk_HaplotypeCaller() {

  SAMPLE_NAME=$(echo $1 | sed -e 's/_sorted_RG_55genes_markdup_BQSR.bam//')
  
  samtools index $1
  
  _JAVA_OPTIONS="-Xmx90g" gatk HaplotypeCaller \
      -R /home/mwkhan/Taste_Genes_Analysis/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
      -I $1 \
      -O sorted_RG_55genes_markdup_BQSR_HaplotypeCaller/${SAMPLE_NAME}_raw_variants.g.vcf.gz \
      --assembly-region-padding 100 \
      -L /home/mwkhan/Taste_Genes_Analysis/References_Seqs/bed_files/Merged_bed_hg38_Jan15_no_names.bed \
      -ERC GVCF
}

export -f process_gatk_HaplotypeCaller
ls    *_sorted_RG_55genes_markdup_BQSR.bam  | parallel -j 20 process_gatk_HaplotypeCaller
## Do HaplotypeCaller on 20 threads
```


