---
title: "Untitled"
output: html_document
date: "2023-08-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
```{r}
rm(list = ls())
```

```{r}
library(here)
library(tidyverse)
```


The CFTR genes coordinates given provided by genome quebec are from grch37/hg19
We need to convert it to grch38
the tool is liftover from ucsc. it also provide a command line tool for the same
Copy the coordinates from the primer fils
paste it to
https://genome.ucsc.edu/cgi-bin/hgLiftOver
change the assemblies option accordingly original: hg19, new: hg38
save the output to CFTR_hg19_to_hg38_from_UCSC_LiftOver.bed
```{r}
CFTR_hg19_to_hg38_from_UCSC_LiftOver = read.table(here("Samples384_CFTR_gene/CFTR_hg19_to_hg38_from_UCSC_LiftOver.bed"),
                                                  )
CFTR_hg19_to_hg38_from_UCSC_LiftOver
```
```{r}
CFTR_hg38_bed = CFTR_hg19_to_hg38_from_UCSC_LiftOver %>% 
  dplyr::select(V1, V2, V3)
CFTR_hg38_bed

write.table(CFTR_hg38_bed,
            here("Samples384_CFTR_gene/CFTR_hg38.bed"),
            col.names = F,
            row.names = F,
            quote = F,
            sep = "\t")

CFTR_hg38_bed_w_gene_name = CFTR_hg38_bed %>% 
  mutate(Gene = "CFTR")
CFTR_hg38_bed_w_gene_name

write.table(CFTR_hg38_bed_w_gene_name,
            here("Samples384_CFTR_gene/CFTR_hg38_bed_w_gene_name.bed"),
            col.names = F,
            row.names = F,
            quote = F,
            sep = "\t")
```
 call gatk HaplotypeCaller on dedup.bam file using cftr bed file

```{bash}
# Activate conda environment
conda activate GATK_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene
mkdir /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files
# Function for processing each file
process_gatk_HaplotypeCaller() {
  #file=$1
  SAMPLE_NAME=$(basename $1 | sed -e 's/_BQSR_dedup_sample_final.bam//')
  echo "Now processsing: "$SAMPLE_NAME
  out_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files
  _JAVA_OPTIONS="-Xmx90g" gatk HaplotypeCaller \
  -R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
  -I $1 \
  -O $out_dir/${SAMPLE_NAME}_raw_variants.g.vcf.gz \
  -bamout $out_dir/${SAMPLE_NAME}_reassamble_sample.out.bam \
  -ERC GVCF \
  --assembly-region-padding 100 \
  -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/CFTR_hg38.bed
}

export -f process_gatk_HaplotypeCaller
input_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/GATK_pipepile/GATK_BS384/FastqToGVCF
# Run the function in parallel for all *.bam files
# ${input_dir}/*006_BQSR_dedup_sample_final.bam ${input_dir}/*008_BQSR_dedup_sample_final.bam  | parallel -j 20 process_gatk_HaplotypeCaller
ls    ${input_dir}/*_BQSR_dedup_sample_final.bam                                               | parallel -j 20 process_gatk_HaplotypeCaller
```

```{bash}
conda activate GATK_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files/
ref_file=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna
# Start the CombineGVCFs command
cmd='_JAVA_OPTIONS="-Xmx90g" gatk CombineGVCFs -R ${ref_file} '
# provide reference file
cmd+= " -L /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/CFTR_hg38.bed"
# Add the output argument
cmd+=" -O Samples384_Combined_CFTR.g.vcf"
# Add each GVCF file to the command
for file in $(ls *_raw_variants.g.vcf.gz); do
  cmd+=" -V $file "
done

# Run the command
eval $cmd
```


```{bash}
conda activate GATK_env

cd  /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files/

_JAVA_OPTIONS="-Xmx90g" gatk GenotypeGVCFs \
-R /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/GCA_000001405.15_GRCh38/GCA_000001405.15_GRCh38_full_analysis_set.fna \
-V ./Samples384_Combined_CFTR.g.vcf \
-O ./Samples384_Combined_CFTR_Genotyped.g.vcf
```

# Remove samples from vcf files
As We do not have metadata for these samples

```{bash}
cd  /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files/
# ./BS_106,./BS_Chrim_25,./BS_Chrim_59,./BS_Chrim_77,./BS_Chrim_79
bcftools view \
-s ^./BS_103,./BS_Chrim_25,./BS_Chrim_59,./BS_Chrim_77,./BS_Chrim_79 \
./Samples384_Combined_CFTR_Genotyped.g.vcf > \
./Samples379_Combined_CFTR_Genotyped.g.vcf
```


```{bash}
grep '^#CHROM' Samples379_Combined_CFTR_Genotyped.g.vcf | cut -f10-
<!-- ./BS_006        ./BS_008        ./BS_014        ./BS_026        ./BS_027         -->
<!-- ./BS_028        ./BS_029        ./BS_031        ./BS_033        ./BS_034         -->
<!-- ./BS_036        ./BS_038        ./BS_040       ./BS_042 -->
<!-- ./BS_043        ./BS_044        ./BS_049        ./BS_050        ./BS_051         -->
<!-- ./BS_089        ./BS_090        ./BS_091        ./BS_092_ii -->
```

## to match the sample names with plink phenotype file we need to remove "./" from all files
so that ./BS_??? become BS???
and "_ii" from some files
```{bash}
cd  /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files/
sed '/^#CHROM/ s/\.\/BS_/BS/g; s/_ii//g' Samples379_Combined_CFTR_Genotyped.g.vcf > \
Samples379_Combined_CFTR_Genotyped_renamed.g.vcf

```


```{bash}
mkdir -p /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis

input_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/GATK_bam_files

bgzip \
-c ${input_dir}/Samples379_Combined_CFTR_Genotyped_renamed.g.vcf > \
Samples379_Combined_CFTR_Genotyped_renamed.g.vcf.gz

tabix \
-p vcf \
Samples379_Combined_CFTR_Genotyped_renamed.g.vcf.gz


bcftools annotate \
-a /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/References_Seqs/Homo_sapiens_assembly38.dbsnp138.vcf.gz \
-c ID \
Samples379_Combined_CFTR_Genotyped_renamed.g.vcf.gz > \
Samples379_Combined_CFTR_Genotyped_renamed_w_snpIDs.g.vcf
```


# VCF default filtering
```{bash}
echo "number of variants before vcftoolsFilt"
bcftools view -H Samples379_Combined_CFTR_Genotyped_renamed_w_snpIDs.g.vcf | wc -l

vcftools \
--vcf Samples379_Combined_CFTR_Genotyped_renamed_w_snpIDs.g.vcf \
--remove-indels \
--recode \
--stdout > Samples379_Combined_CFTR_w_snpIDs_vcftoolsFilt.g.vcf

echo "number of variants after vcftoolsFilt"
bcftools view -H Samples379_Combined_CFTR_w_snpIDs_vcftoolsFilt.g.vcf | wc -l
```


```{bash}
conda activate Plink_env
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis
metadata_dir=/media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Metadata_files

plink \
--vcf ./Samples379_Combined_CFTR_w_snpIDs_vcftoolsFilt.g.vcf \
--double-id \
--make-bed \
--keep-allele-order \
--make-pheno ${metadata_dir}/Samples379_Combined_Phenotype.txt 1 \
--update-sex ${metadata_dir}/Samples379_Combined_Phenotype_with_sex_info.txt \
--recode \
--out ./Samples379_Combined_CFTR_Plink
```


```{bash}
conda activate Plink_env

cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis
plink \
--file ./Samples379_Combined_CFTR_Plink \
--geno 0.2 \
--maf 0.01 \
--not-chr x \
--hwe 0.001 \
--recode vcf \
--assoc \
--allow-no-sex \
--out ./Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001
```

```{bash}
cd /media/Second_stor/wasif/WK_Rprojects/RP_230718_Taste_Genes_Analysis_1/Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis
conda activate SnpEff_env

snpEff \
-v hg38 \
./Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001.vcf > \
./Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001_anno.vcf
```
```{r}

# Read VCF file
vcf <- read.vcfR(here("Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis/Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001_anno.vcf"))


# Extract chromosome and position
chrom <- vcf@fix[,1]
pos <- vcf@fix[,2]

# Extract the INFO column
info <- vcf@fix[,8]

# Extract the ANN information from INFO by removing the 'ANN=' prefix
ann_info <- gsub("^ANN=", "", info)

# Each element of 'ann_info' can have multiple annotations (separated by comma),
# and each annotation has multiple sub-fields (separated by '|').

# Function to parse a single annotation
parse_ann <- function(ann) {
  # Split the annotation into sub-fields
  sub_fields <- strsplit(ann, "\\|")[[1]]
  
  # Return the gene name and variant type if they exist
  if (length(sub_fields) >= 4) {
    return(c(sub_fields[4], sub_fields[2])) # sub_fields[2] contains the variant type
  } else {
    return(c(NA, NA))
  }
}

# Parse the ANN field for each variant
genes_and_variant_types <- t(sapply(ann_info, parse_ann)) # transpose to make each gene-variant pair a row

# Create a dataframe with the chromosome, position, gene name, and variant type
df_chromPOS_to_gene_variant <- data.frame(CHROM = chrom, POS = pos, GENE = genes_and_variant_types[,1], VARIANT_TYPE = genes_and_variant_types[,2])
df_chromPOS_to_gene_variant
```
## Assign gene name to assoc file
```{r}
assoc_file = read.table(here("Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis/Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001.assoc"),
                        header = T)
assoc_file
assoc_file = assoc_file %>% 
  mutate(BP = as.character(BP),
         CHR = as.character(CHR))
  
assoc_file
```

# merge
```{r}
assoc_file_df_chromPOS_to_gene_variant =  inner_join(assoc_file, df_chromPOS_to_gene_variant,
                                             by = c("CHR" = "CHROM", "BP" = "POS"))
assoc_file_df_chromPOS_to_gene_variant

assoc_file_df_chromPOS_to_gene_variant_adjP = assoc_file_df_chromPOS_to_gene_variant %>% 
  mutate(p_BH = p.adjust(P, method = "BH"),
         p_bonferroni = p.adjust(P, method = "bonferroni") ) %>% 
  arrange(p_bonferroni) %>% 
  mutate(BP= as.numeric(BP))
assoc_file_df_chromPOS_to_gene_variant_adjP

write.table(assoc_file_df_chromPOS_to_gene_variant_adjP,
            here("Samples384_Analysis/Samples384_CFTR_gene/Plink_analysis/Samples379_Combined_CFTR_Plink_filtered_geno.2_maf.01_hwe.001_assoc_w_GENE.csv"),
            row.names = F,
            quote = F,
            sep = ",")
```